<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Adam J Taylor" />
  <meta name="author" content="Dimitrios Tsikritsis" />
  <meta name="author" content="Alex Dexter" />
  <meta name="author" content="Amy Burton" />
  <meta name="author" content="Josephine Bunch" />
  <meta name="author" content="Natalie Belsey" />
  <meta name="dcterms.date" content="2021-10-15" />
  <meta name="keywords" content="desorption electrospray ionization mass spectrometry, transmission Raman spectroscopy, solid oral dosage form, classification" />
  <title>Complementary classification of solid oral dosage forms in ambient conditions by desorption electrospray ionization mass spectrometry and transmission Raman spectroscopy</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/main/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="Complementary classification of solid oral dosage forms in ambient conditions by desorption electrospray ionization mass spectrometry and transmission Raman spectroscopy" />
  <meta name="citation_title" content="Complementary classification of solid oral dosage forms in ambient conditions by desorption electrospray ionization mass spectrometry and transmission Raman spectroscopy" />
  <meta property="og:title" content="Complementary classification of solid oral dosage forms in ambient conditions by desorption electrospray ionization mass spectrometry and transmission Raman spectroscopy" />
  <meta property="twitter:title" content="Complementary classification of solid oral dosage forms in ambient conditions by desorption electrospray ionization mass spectrometry and transmission Raman spectroscopy" />
  <meta name="dc.date" content="2021-10-15" />
  <meta name="citation_publication_date" content="2021-10-15" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="Adam J Taylor" />
  <meta name="citation_author_institution" content="National Physical Laboratory, Teddington, UK" />
  <meta name="citation_author_institution" content="Current affiliation: Sage Bionetworks, Seattle, WA, USA" />
  <meta name="citation_author" content="Dimitrios Tsikritsis" />
  <meta name="citation_author_institution" content="National Physical Laboratory, Teddington, UK" />
  <meta name="citation_author" content="Alex Dexter" />
  <meta name="citation_author_institution" content="National Physical Laboratory, Teddington, UK" />
  <meta name="citation_author" content="Amy Burton" />
  <meta name="citation_author_institution" content="National Physical Laboratory, Teddington, UK" />
  <meta name="citation_author_institution" content="Current affiliation: GlaxoSmithKline, Stevenage, UK" />
  <meta name="citation_author" content="Josephine Bunch" />
  <meta name="citation_author_institution" content="National Physical Laboratory, Teddington, UK" />
  <meta name="citation_author_institution" content="Imperial College London, London, UK" />
  <meta name="citation_author_institution" content="The Rosalind Frankilin Insitute, Harwell, UK" />
  <meta name="citation_author" content="Natalie Belsey" />
  <meta name="citation_author_institution" content="National Physical Laboratory, Teddington, UK" />
  <link rel="canonical" href="https://adamjtaylor.github.io/tablet-classification/" />
  <meta property="og:url" content="https://adamjtaylor.github.io/tablet-classification/" />
  <meta property="twitter:url" content="https://adamjtaylor.github.io/tablet-classification/" />
  <meta name="citation_fulltext_html_url" content="https://adamjtaylor.github.io/tablet-classification/" />
  <meta name="citation_pdf_url" content="https://adamjtaylor.github.io/tablet-classification/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://adamjtaylor.github.io/tablet-classification/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://adamjtaylor.github.io/tablet-classification/v/1b89ec171a2469a220f3f6acc5222e26e9925f49/" />
  <meta name="manubot_html_url_versioned" content="https://adamjtaylor.github.io/tablet-classification/v/1b89ec171a2469a220f3f6acc5222e26e9925f49/" />
  <meta name="manubot_pdf_url_versioned" content="https://adamjtaylor.github.io/tablet-classification/v/1b89ec171a2469a220f3f6acc5222e26e9925f49/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Complementary classification of solid oral dosage forms in ambient conditions by desorption electrospray ionization mass spectrometry and transmission Raman spectroscopy</h1>
</header>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>Adam J Taylor</strong><br><br>
<small>
National Physical Laboratory, Teddington, UK; Current affiliation: Sage Bionetworks, Seattle, WA, USA
</small></p></li>
<li><p><strong>Dimitrios Tsikritsis</strong><br><br>
<small>
National Physical Laboratory, Teddington, UK
</small></p></li>
<li><p><strong>Alex Dexter</strong><br><br>
<small>
National Physical Laboratory, Teddington, UK
</small></p></li>
<li><p><strong>Amy Burton</strong><br><br>
<small>
National Physical Laboratory, Teddington, UK; Current affiliation: GlaxoSmithKline, Stevenage, UK
</small></p></li>
<li><p><strong>Josephine Bunch</strong><br><br>
<small>
National Physical Laboratory, Teddington, UK; Imperial College London, London, UK; The Rosalind Frankilin Insitute, Harwell, UK
</small></p></li>
<li><p><strong>Natalie Belsey</strong><br><br>
<small>
National Physical Laboratory, Teddington, UK
</small></p></li>
</ul>
<h2 class="page_break_before" id="abstract">Abstract</h2>
<p>Discrepancies or defects in active ingredients, excipients and coatings that form solid oral dosage forms can both impact product quality and provide hallmarks of off-brand or counterfeit products. There is therefore a need for rapid and continuous analytical techniques that can assess and classify product differences of intact samples at- or near the production line, or in analytical labs, ideally without resorting to product dissolution.</p>
<p>Here we test the ability of two rapid ambient chemical characterization methods to discriminate between solid dosage forms: desorption electrospray ionization mass spectrometry and transmission Raman spectroscopy. These two techniques are highly complementary, offering greater sensitivity to the analysis of the surface and the tablet bulk, respectively. The data sets generated were then used to test a variety of classification algorithms including linear discriminate analysis, tree-based methods, a simple neural network, and support vector machines (SVM). The highest performing algorithms for DESI-MSI were the SVM, with an additional performance boost when used with a polynomial kernel. For transmission Raman data, a linear discriminant analysis (LDA) model was found to be the most effective.</p>
<h2 id="introduction">Introduction</h2>
<p>Inconsistencies in active ingredients, excipients, the thickness and integrity of coatings and the presence of impurities in solid oral dosage forms all negatively affect their performance.
Inferior quality attributes can be useful to identify off-brand or counterfeit products.
There is a need for rapid and continuous analytical techniques that can assess and classify product differences of intact samples at- or near the production line, or in analytical labs, ideally without resorting to product dissolution<span class="citation" data-cites="Vu1f6wmd"><sup><a href="#ref-Vu1f6wmd" role="doc-biblioref">1</a></sup></span>.
Rapid measurement tools are particularly important to enable continuous monitoring, necessary to support the change from batch to continuous manufacturing.
Analytical methods are required to monitor both the actives, coatings and consistency of the product: For example, in addition to the total API content, insight is also needed on degradation products, impurities, (co-) crystallinity/presence of polymorphs, and content uniformity.
The ability to monitor tablet coating thickness and integrity is of great importance, particularly for functional coatings, such gastro-resistance, which would be compromised by insufficient thickness, or the occurrence of cracks in the film<span class="citation" data-cites="9XSo86vE"><sup><a href="#ref-9XSo86vE" role="doc-biblioref">2</a></sup></span>.</p>
<p>Quantitative analysis of pharmaceutical tablets is routinely performed by HPLC which offers accurate and sensitive measurements of the active ingredient(s) and excipients, in addition to the presence of any contaminants.
However, solution-based analytical methods are destructive and labor-intensive.</p>
<p>Mass spectrometric methods can provide unlabeled identification, both of expected ingredients in known samples and of contaminants or components of unknown formulations.
Ambient ionization mass spectrometry approaches including DESI (desorption electrospray ionisation) and DART (direct analysis in real time) facilitate the desorption and ionization from the surface of samples at atmospheric conditions, without dissolution or additional sample preparation.
They are therefore potentially useful tools for rapid assessment of solid oral dosage forms</p>
<p>Optical spectroscopy techniques offer rapid, non-destructive analysis, including polymorphic identification<span class="citation" data-cites="8hsFY7pn"><sup><a href="#ref-8hsFY7pn" role="doc-biblioref">3</a></sup></span>, and are also able to measure insoluble ingredients.
They have consequently been exploited for in-line process analytical testing and as quality control tools<span class="citation" data-cites="Vu1f6wmd"><sup><a href="#ref-Vu1f6wmd" role="doc-biblioref">1</a></sup></span>.
For example, infra-red-based techniques (FTIR and DESI<span class="citation" data-cites="1GTjZYdpk"><sup><a href="#ref-1GTjZYdpk" role="doc-biblioref">4</a></sup></span>; and NIR and Mid-IR spectroscopy classification of MDMA containing tablets<span class="citation" data-cites="1ElQCA6hC"><sup><a href="#ref-1ElQCA6hC" role="doc-biblioref">5</a></sup></span>.
Near infrared is the most commonly used process analytical tool<span class="citation" data-cites="Vu1f6wmd"><sup><a href="#ref-Vu1f6wmd" role="doc-biblioref">1</a></sup></span>, however Raman spectroscopy provides complementary information and has grown in popularity in recent years, since it provides more distinct spectral features, and is better-suited to analysis in aqueous environments owing to the relatively weak strength of the Raman O-H band.
Technological advancements have facilitated miniaturization, increased speed and reduced cost, resulting in more widespread implementation.<span class="citation" data-cites="YUItDAnD"><sup><a href="#ref-YUItDAnD" role="doc-biblioref">6</a></sup></span></p>
<h3 id="ambient-ionisation-mass-spectrometry-of-tablets">Ambient ionisation mass spectrometry of tablets</h3>
<p>Desorption electrospray ionization uses a charged electrospray of organic solvent which, when directed at the sample surface in proximity to the mass spectrometry inlet, desorbs ions from the sample which may be taken up into a mass spectrometer<span class="citation" data-cites="8OEKQBYr"><sup><a href="#ref-8OEKQBYr" role="doc-biblioref">7</a></sup></span>.
As this process takes place at ambient pressure and with a flexible geometry, the technique is suited for the analysis of a wide range of samples including explosives on surfaces<span class="citation" data-cites="nGY6XRTE"><sup><a href="#ref-nGY6XRTE" role="doc-biblioref">8</a></sup></span>, fingerprints<span class="citation" data-cites="1A0sMbTaa"><sup><a href="#ref-1A0sMbTaa" role="doc-biblioref">9</a></sup></span>, plants<span class="citation" data-cites="tvTiSD66"><sup><a href="#ref-tvTiSD66" role="doc-biblioref">10</a></sup></span> and tissues<span class="citation" data-cites="sLghpFT8 1ClzsqfDW"><sup><a href="#ref-sLghpFT8" role="doc-biblioref">11</a>,<a href="#ref-1ClzsqfDW" role="doc-biblioref">12</a></sup></span></p>
<p>One of the early descriptions of DESI-MSI was in the profiling of tablets<span class="citation" data-cites="1EVpL1egt"><sup><a href="#ref-1EVpL1egt" role="doc-biblioref">13</a></sup></span>.
Chen et al demonstrated the use of DESI-MS to profile tablets containing loratadine, folic acid, acetaminophen (paracetamol), aspirin, melatonin or caffeine.
Optimization of DESI parameters including voltage, solvent delivery and capillary temperature facilitated analysis at up to three scans per second.<br />
Subsequent studies using DESI-MS of tablets have focused on targeted analysis for active ingredients. For example, the identification MDMA and amphetamine derivatives in ecstasy tablets<span class="citation" data-cites="or5TqRX6"><sup><a href="#ref-or5TqRX6" role="doc-biblioref">14</a></sup></span>, counterfeit artesunate antimalarial tablets<span class="citation" data-cites="1GTjZYdpk Cc9lNkvZ"><sup><a href="#ref-1GTjZYdpk" role="doc-biblioref">4</a>,<a href="#ref-Cc9lNkvZ" role="doc-biblioref">15</a></sup></span> and antiviral capsules<span class="citation" data-cites="4iiQhDKH"><sup><a href="#ref-4iiQhDKH" role="doc-biblioref">16</a></sup></span>.</p>
<p>For ambient mass spectrometry to be deployable in the field for counterfeiting applications, or in manufacturing environments for QA/QC, the mass spectrometer must be compact.
Several designs for small field-deployable mass spectrometers have been demonstrated with DESI MS sources<span class="citation" data-cites="1Gg9u8kt2 UuPy88nX"><sup><a href="#ref-1Gg9u8kt2" role="doc-biblioref">17</a>,<a href="#ref-UuPy88nX" role="doc-biblioref">18</a></sup></span>.</p>
<p>Each of these applications has targeted expected components of the tablet of interest, predominantly active ingredients or excipients.
However, in manufacturing QA/QC and counterfeit-detection applications, additional information on unexpected changes in active or excipient source or quality, as well as the introduction of contaminants may be of importance.
Untargeted multivariate and machine learning approaches are therefore of interest to determine differences between samples using all spectral information.</p>
<p>Classification approaches for mass spectrometry applications are proving powerful in a range of applications. The two most widespread applications of classification in mass spectrometry are in disease diagnosis and determination of bacterial type<span class="citation" data-cites="TuCEd7AU"><sup><a href="#ref-TuCEd7AU" role="doc-biblioref">19</a></sup></span>.
A range of classification algorithms have been applied to mass spectrometry and spectroscopy data. PLS-DA is most commonly reported, although a range of algorithms including neural networks, and support vector machines<span class="citation" data-cites="e0eRY8v5"><sup><a href="#ref-e0eRY8v5" role="doc-biblioref">20</a></sup></span> have been reported.
Several publications have evaluated different classification algorithms but unsurprisingly the optimal algorithm depends greatly on the nature of the input data. A summary of classification and other data analysis for proteomics can be found here<span class="citation" data-cites="ydPgvUPy"><sup><a href="#ref-ydPgvUPy" role="doc-biblioref">21</a></sup></span> and for metabolomics here<span class="citation" data-cites="qcR4mVsb"><sup><a href="#ref-qcR4mVsb" role="doc-biblioref">22</a></sup></span>.
Classification approaches are becoming more accessible through modeling tools with consistent grammar and data structure, and their integration into mass spectrometry software [ScilsLab,Waters software]<span class="citation" data-cites="qcR4mVsb"><sup><a href="#ref-qcR4mVsb" role="doc-biblioref">22</a></sup></span>.</p>
<p>Notably, classification of rapid evaporative ionization MS enables real-time classification of tissue types during surgery<span class="citation" data-cites="O5djQ3nR"><sup><a href="#ref-O5djQ3nR" role="doc-biblioref">23</a></sup></span>.
Classification of REIMS data has also found applications in food security<span class="citation" data-cites="g08cPiPx"><sup><a href="#ref-g08cPiPx" role="doc-biblioref">24</a></sup></span> and bacterial speciation<span class="citation" data-cites="fKTaMhqM"><sup><a href="#ref-fKTaMhqM" role="doc-biblioref">25</a></sup></span>.
Classification approaches have also been widely employed in mass spectrometry imaging data, particularly in the classification of cancerous tissue<span class="citation" data-cites="1Dw5WtMxe"><sup><a href="#ref-1Dw5WtMxe" role="doc-biblioref">26</a></sup></span></p>
<h3 id="raman-spectroscopy-analysis-of-tablets">Raman spectroscopy analysis of tablets</h3>
<p>Raman spectroscopy exploits the inelastic scattering of light by the sample to reveal valuable chemical and structural information. Information can be obtained from the sample in a non-destructive manner, making it a popular process analytical technology tool.
Raman spectroscopy can be performed in a variety of sampling configurations/geometries, the most appropriate depending on the application.
Confocal Raman microscopy can provide detailed chemical mapping with high spatial resolution, however this is generally reserved for forensic investigation rather than continuous monitoring, since it requires lengthy acquisition times.
Sub-sampling issues associated with conventional backscattered Raman can be overcome by strategies such as sample rotation in conjunction with spectral averaging, or simultaneous wide angle illumination<span class="citation" data-cites="jh1jwVUX"><sup><a href="#ref-jh1jwVUX" role="doc-biblioref">27</a></sup></span>.</p>
<p>Matousek et al demonstrated the ability of transmission Raman spectroscopy to probe deep into turbid materials such as pharmaceutical tablets and provide information on their bulk properties<span class="citation" data-cites="13x0WMYJq vvUUyBtN"><sup><a href="#ref-13x0WMYJq" role="doc-biblioref">28</a>,<a href="#ref-vvUUyBtN" role="doc-biblioref">29</a></sup></span>.
In contrast to conventional backscattered Raman, in transmission Raman spectroscopy the beam passes through the full thickness of the tablet, sampling a much larger volume of the material, and consequently provides more representative sampling<span class="citation" data-cites="eL1MOkeG"><sup><a href="#ref-eL1MOkeG" role="doc-biblioref">30</a></sup></span>.
Although Raman scattering intensity is linear with concentration within the same confocal plane, transmission Raman signal intensity is slightly biased towards the bulk of the tablet relative to the exterior due to internal scattering<span class="citation" data-cites="Qy6BdVBn"><sup><a href="#ref-Qy6BdVBn" role="doc-biblioref">31</a></sup></span>.
In contrast, DESI-MSI sampling is biased towards the surface/coating composition. Therefore, in combination, these two techniques should provide a powerful toolkit with which to assess compositional differences between pharmaceutical tablet formulations.</p>
<p>Raman spectra of complex mixtures such as solid dosage forms often have complicated spectra with overlapping peaks.
For this reason, multivariate techniques are often applied to help identify the components of interest and changes in chemistry.
The selection and use of unsupervised and/or supervised techniques on Raman spectra rely on factors such as prior knowledge of the raw component spectra, and the quantity and complexity of the spectra<span class="citation" data-cites="1EsJaNvp"><sup><a href="#ref-1EsJaNvp" role="doc-biblioref">32</a></sup></span>.</p>
<p>As with mass spectrometry, classification of Raman spectroscopy data has been primarily focused on disease diagnostics<span class="citation" data-cites="ubp6jD46 koyIoA0l hmGuG8Yj"><sup><a href="#ref-ubp6jD46" role="doc-biblioref">33</a>–<a href="#ref-hmGuG8Yj" role="doc-biblioref">35</a></sup></span> and bacterial analysis<span class="citation" data-cites="orJZ0nNQ 14piFxIYf"><sup><a href="#ref-orJZ0nNQ" role="doc-biblioref">36</a>,<a href="#ref-14piFxIYf" role="doc-biblioref">37</a></sup></span>. Other noteworthy examples of the use of classification in Raman spectroscopy include differentiation of narcotics<span class="citation" data-cites="W0VneN3Y"><sup><a href="#ref-W0VneN3Y" role="doc-biblioref">38</a></sup></span>, pharmaceuticals<span class="citation" data-cites="6wD6szWs"><sup><a href="#ref-6wD6szWs" role="doc-biblioref">39</a></sup></span>,<span class="citation" data-cites="d5haWH7e"><sup><a href="#ref-d5haWH7e" role="doc-biblioref">40</a></sup></span>, and counterfeit tablets<span class="citation" data-cites="TToYRrJQ"><sup><a href="#ref-TToYRrJQ" role="doc-biblioref">41</a></sup></span>.</p>
<p>There have been relatively few comparisons of different classification methods for Raman spectroscopy data.
Zheng et al. compared SVM, LDA and k-nearest neighbours (KNN) methods to classify renin hypertension from Raman data from serum<span class="citation" data-cites="16nSdQZQl"><sup><a href="#ref-16nSdQZQl" role="doc-biblioref">42</a></sup></span>.
They found that SVM and LDA performed similarly, and both outperformed the KNN algorithm. Partial least squares (PLS) and PLS discriminant analysis are also commonly used methods in characterizing tablets, however care is required depending on the data quantity and the pre-preprocessing performed<span class="citation" data-cites="5Jy2Axu0"><sup><a href="#ref-5Jy2Axu0" role="doc-biblioref">43</a></sup></span>.
Qun et al. tested the classification of expired drugs using PLS-DA, SVM and KNN, and reported that SVM gave the strongest performance<span class="citation" data-cites="1FNJpTdZe"><sup><a href="#ref-1FNJpTdZe" role="doc-biblioref">44</a></sup></span>. F
ransson et al tested the performance of multivariate methods including PLS, classical least squares (CLS) and multivariate curve resolution (MCR) for classification of pharmaceutical tablets<span class="citation" data-cites="BmWj6JSS"><sup><a href="#ref-BmWj6JSS" role="doc-biblioref">45</a></sup></span>.</p>
<h3 id="objective">Objective</h3>
<p>In this study we set out to explore the potential of DESI-MSI and transmission Raman spectroscopy to distinguish commercially available pharmaceutical tablets with similar or different formulations.
Pairing DESI with transmission Raman spectroscopy was of particular interest due to their complementarity and relative abilities to sensitively probe the surface vs the bulk of the tablets.
Classification of tablets based on both active ingredients and excipients has the potential to be used for in-line quality control measures during pharmaceutical manufacturing, and for rapid counterfeit testing.
As such we have tested a range of classification algorithms on their capability to differentiate these tablets using a range of pre-processing methods to determine the best approaches to use in different applications.</p>
<h2 id="experimnental">Experimnental</h2>
<h3 id="samples">Samples</h3>
<p>Samples were selected from commercially available of-the-shelf products and purchased from a local supplier. Their names, active ingredients, listed excipients and MHRA product license numbers are included in Table <a href="#tbl:samples">1</a></p>
<div id="tbl:samples" class="tablenos">
<table id="tbl:samples">
<caption><span>Table 1:</span> Details of the tablet types analysed. Letter codes for each type are used throughout
</caption>
<colgroup>
<col style="width: 8%" />
<col style="width: 19%" />
<col style="width: 27%" />
<col style="width: 25%" />
<col style="width: 19%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Type</th>
<th style="text-align: center;">Product name</th>
<th style="text-align: center;">Active ingredients</th>
<th style="text-align: center;">Listed excipients</th>
<th style="text-align: center;">MHRA licence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">A</td>
<td style="text-align: center;">Anadin Extra Tablets</td>
<td style="text-align: center;">300 mg Aspirin, 200 mg Paracetamol, 45 mg Caffeine</td>
<td style="text-align: center;">Maize starch, microcrystalline cellulose (E460), hydrogenated vegetable oil, hydroxypropyl methylcellulose (E464), polyethylene glycol, pregelatinised starch and povidone</td>
<td style="text-align: center;">PL 00165/5013R</td>
</tr>
<tr class="even">
<td style="text-align: left;">B</td>
<td style="text-align: center;">Tesco Paracetamol Extra Tablets</td>
<td style="text-align: center;">500 mg Paracetamol, 65 mg Caffeine</td>
<td style="text-align: center;">tarch, povidone k-30, povidone k-90, croscarmellose sodium, talc, stearic acid and magnesium stearate</td>
<td style="text-align: center;">PL 08977/0025</td>
</tr>
<tr class="odd">
<td style="text-align: left;">C</td>
<td style="text-align: center;">Tesco Paracetamol Tablets</td>
<td style="text-align: center;">500 mg Paracetamol</td>
<td style="text-align: center;">Potato Starch, pregelatinised starch, magnesium stearate, povidone, stearic acid and talc</td>
<td style="text-align: center;">PL 08977/0014</td>
</tr>
<tr class="even">
<td style="text-align: left;">D</td>
<td style="text-align: center;">Tesco Extra Power Pain Control Tablets</td>
<td style="text-align: center;">300 mg Aspirin, 200 mg Paracetamol, 45 mg Caffeine</td>
<td style="text-align: center;">Povidone, hydroxypropylcellulose, stearic acid, microcrystalline cellulose, maize starch, pregelatinised starch, hydroxypropyl methylcellulose 5cPs, hydroxypropyl methylcellulose 15cPs, macrogol 4000</td>
<td style="text-align: center;">PL 29831/0164</td>
</tr>
</tbody>
</table>
</div>
<h3 id="desi-ms">DESI MS</h3>
<p>DESI-MS measurements were performed on a Synapt G2-Si Q-IM-ToF mass spectrometer (Waters Corp). The instrument was operated in ‘resolution mode’. The ion mobility cell was not used. Positive ion mode spectra were collected with a scan time of 1 second across a mass range of m/z 50 to m/z 1200. The instrument was fitted with a prototype DESI source (Waters Corp), with the sprayer configured for electroflow focusing with a fused silica capillary sitting approximately 1 mm behind a 200 µm steel orifice. Methanol with 5% water by volume was delivered at 2 µl/m by a pressure pump (Dolomite). Nitrogen gas was delivered at 0.2 MPa. The spray voltage was set at 5 kV. A heated inlet capillary was set to a calibrated temperature of 400°C using a PID (Waters Research Centre, Hungary). Tablets were sampled by holding the tablet 1-2 mm away from the DESI spray head using plastic tweezers. For training data, acquisition was started with the tablet already under the spray head, such that only data from the tablet surface was acquired, while validation data was collected continuously.</p>
<h3 id="transmission-raman-spectroscopy">Transmission Raman spectroscopy</h3>
<p>Transmission Raman spectra were acquired using a Renishaw InVia Qontor Raman microscope equipped with a 830 nm excitation source fibre-coupled to an InVia transmission Raman accessory (Renishaw plc, Wotton-under-Edge, Gloucestershire), in a temperature controlled environment. Light was collected in transmission with the x5 air objective lens (0.12 NA, N-PLAN, Leica). Tablets were carefully placed onto a flat silicon sample support with a hole just smaller than the tablet dimensions, so that the excitation beam was able to pass through the tablet but not the sample support. Six tablets were analysed of each type in pseudo-random order. Three measurement replicates were acquired, each complete data set was collected on three separate days.</p>
<p>For all tablets, extended spectra were acquired using Renishaw Wire (version 5.3) software for the spectral range of 50 to 1800 cm<sup>-1</sup>, with an acquisition time of 30 seconds, and 5 accumulations. Laser power was set to 100% which has been measured at the sample to be approximately 117 mW. An internal silicon calibration reference spectrum was acquired each day to correct the Raman shift of the data.</p>
<h3 id="data-analysis">Data analysis</h3>
<p>All data were analyzed in R version 3.6.2 (2019-12-12) “Dark and Stormy Night” and RStudio Server version 1.2.5019. Analysis was conducted using the tidyverse<span class="citation" data-cites="EWll2pyF"><sup><a href="#ref-EWll2pyF" role="doc-biblioref">46</a></sup></span> and tidymodels<span class="citation" data-cites="4zds2yP"><sup><a href="#ref-4zds2yP" role="doc-biblioref">47</a></sup></span> metapackages. Raman data preprocessing was conducted in MATLAB 2020a. All analysis was performed on a Linux workstation (Intel Core i9-7900X CPU with 10 cores @ 3.30 GHz, 128G RAM, Ubuntu 16.04.6 LTS.</p>
<h3 id="desi-preprocessing">DESI preprocessing</h3>
<p>For model development and comparison, data were converted from Waters raw format to mzML format using ProteoWizard MSConvert version 3.0.19239-0ae547798. These were read into R using the mzR package. All spectra were re-binned onto the same mass axis with a bin width of m/z 0.01. A mean spectrum of all training data was peak picked using the findPeaks function from the prcma package with a peak intensity threshold of three times the median intensity of the spectrum. 1217 peaks were found. Each spectrum was then individually integrated across the found peak widths to form a datacube. Each scan of the validation dataset was similarly integrated across the peak widths from the training dataset. Reduced peak datacubes were generated by filtering for the top n most intense peaks. Down-binning to simulate reduced mass resolving power was performed by rounding m/z values and summing intensities within each rounded m/z bin.</p>
<h3 id="transmission-raman-spectroscopy-preprocessing">Transmission Raman spectroscopy preprocessing</h3>
<p>Cosmic ray removal was performed automatically by Renishaw Wire (version 5.3) and spectra exported to .txt format. The Raman spectra were baseline corrected using the msbackadj() Matlab function. The baseline was estimated within multiple shifted windows of width 20 separation units, then a spline approximation was used to regress the varying baseline to the window points. The estimated baseline for each spectrum was then subtracted from the corresponding original. The background subtracted spectra were read in R for subsequent processing and analysis. The data were normalized to total spectrum intensity and the Raman shift recalibrated using the weighted-mean centroid to the 520.7 cm-1 peak from the daily Si wafer sample spectrum as a reference. Extended spectra were truncated to a wavenumber range between 250 cm-1 and 1700 cm-1. Due to the limited number of wavenumber bins and the challenges of peak-picking Raman data, the continuous data were taken forward for classification.</p>
<h3 id="classification">Classification</h3>
<p>Spectra were collated into a 10-fold cross validation 1 with 10 repeats in a 4/1/5 (train/test/total) split. To remove highly co-variate polymer peak sequences leading to overfitting, highly correlating variables (Pearson correlation &gt; 0.9) were removed from DESI MSI data. Data were centered around the arithmetic mean and scaled to have a standard deviation of one. Underrepresented classes (for DESI MSI, the background class) were up sampled. Each training fold was applied to a range of classification algorithms using the tidymodels package. All models were implemented with their default parameters beyond setting to classification mode. The functions, engines and default parameters used for each model are provided in supplementary table 1. These models were then used to predict each testing fold. For each fold the F1 score was calculated. For DESI MSI the algorithm with the highest F1 score, a support vector machine with a polynomial kernel was selected for further model tuning on a single 4/1/5 validation split of the training data. For transmission Raman data a LDA model was selected. A final model was fitted on all the training data. These models were used to predict the test independent test sets. Cosine similarity between spectra were calculated using the cosine function from the coop package<span class="citation" data-cites="WsAj0wOM"><sup><a href="#ref-WsAj0wOM" role="doc-biblioref">48</a>, =coop</sup></span>. Considering the angle between vectors, rather than magnitude, cosine similarity provides a useful and robust measure of spectral similarity for highly multivariate datasets<span class="citation" data-cites="XgRF2vnB"><sup><a href="#ref-XgRF2vnB" role="doc-biblioref">49</a></sup></span></p>
<h2 id="results-and-discussion">Results and discussion</h2>
<h3 id="acquisition-of-desi-ms-spectra-from-tablets">Acquisition of DESI-MS spectra from tablets</h3>
<p>Rich mass spectra were rapidly obtained from each tablet type when held under the DESI sprayer and MS inlet capillary (Figure 1A). DESI-MS spectra from tablet types A and D are dominated by repeating polymeric peaks above <em>m/z</em> ~700, indicating the presence of a polymer film coating while spectra from tablet types B and C are less complex, being dominated by peaks below <em>m/z</em> 700. The maximum intensity of the spectra collected in the absence of a tablet (“background”) is lower than for spectra where a tablet is presence.</p>
<p>The polymeric peak sequences observed in tablet type A and D above <em>m/z</em> 700 are distinct from one another with several different peak sequences are observed (Figure 1B). Tablet types A and D both exhibit a clear sequence of peaks spaced by <em>m/z</em> ~44 the a signally charged unit difference from [C<sub>2</sub>H<sub>4</sub>O]<sup>+</sup>. For both tablet types peaks above <em>m/z</em> 1000 are most intense. In tablet type A this predominantly consists of isotope clusters separated by <em>m/z</em> ~14.68, with peaks separated by <em>m/z</em> 0.33, indicating a 3+ charge state of the loss of [C2H4O]. Conversely, in tablet type D the isotope clusters are separated by <em>m/z</em> ~11.01 with peak spacing of <em>m/z</em> ~0.25, indicating a 4+ charge state and the loss of [C<sub>2</sub>H<sub>4</sub>O]. This indicates that while both are coated with a PEG-based polymer, the molecular weight or coating application may differ between the brand name and generic product.</p>
<p>All three active ingredients were annotated from a mean spectrum within 15 ppm mass accuracy. Boxplots of single scan intensity show distinct differences in active intensity between types (Figure 1C, 1D, 1E). Caffeine is present in tablet types A, B and D and absent from type C. Median intensity is highest for the uncoated tablet type B. Intensity for tablet type C is similar to that in the background. While aspirin is present in tablet types A and D, the detected intensity in type B is high. This may represent an isobaric compound also present in type B or carry over from sampling tablet type A. Paracetamol is present in, and detected in all table types, although is most intense in tablets without a film-coating (types B and C) which may otherwise mask signal from the underlying bulk material. Variance for all actives and tablet types was high, indicating the need for consistent sampling procedures and robust classification approaches.</p>
<h3 id="acquisition-of-transmission-raman-spectra-from-tablets">Acquisition of Transmission Raman spectra from tablets</h3>
<p>Mean transmission Raman spectra for each tablet type (Figure 1F) show very similar profiles for tablet types A and D (the film coated aspirin, paracetamol, and caffeine tablets) and between types B (paracetamol and caffeine and C (paracetamol only). Peaks characteristic of each active ingredient are observed (Caffeine: 555 cm<sup>-1</sup>, Paracetamol: 858 cm<sup>-1</sup>, Aspirin: 1192 cm<sup>-1</sup>)<span class="citation" data-cites="1BDd0dQq8 148KCDHV1 BoUXYyf5"><sup><a href="#ref-1BDd0dQq8" role="doc-biblioref">50</a>–<a href="#ref-BoUXYyf5" role="doc-biblioref">52</a></sup></span>. However, no clear differences between tablets A and D were observed. As expected based on the tablet composition, the predominant spectral features relate to the actives rather than from the coating ingredients.</p>
<div id="fig:spectra" class="fignos">
<figure>
<img src="https://github.com/adamjtaylor/tablet-classification/raw/main/content/figures/fig1.png" alt="Figure 1: Representative spectra from DESI MS and transmission Raman spectroscopy analysis of solid-oral dosage forms (A-C) DESI MS results. (A) Mean spectra for tablet types A, B, C &amp; D and background. (B) Mean spectra for tablet types A (teal) and D (pink) for mass range m/z 1065 to m/z 1090. (C-E) Boxplots for peaks assigned as caffeine (C), aspirin (D) and paracetamol (E) for each tablet type and background. (F-L) Transmission Raman spectroscopy results. (F) Mean spectra for tablet type A-D from training data. Peaks for the active ingredients shown in G-L are highlighted with a vertical line. (G-L) Mean spectra and (J-L) boxplots for peaks annotated as active ingredients (G &amp; J) Caffeine, 541 cm-1, (H &amp; K) Paracetamol, 847 cm-1, and (I &amp; L) Aspirin, 1181 cm-1). Boxplots show median (horizontal line), interquartile range (box) and range excluding outliers (whiskers). Points show single scan intensities." /><figcaption aria-hidden="true"><span>Figure 1:</span> <strong>Representative spectra from DESI MS and transmission Raman spectroscopy analysis of solid-oral dosage forms</strong>
(A-C) DESI MS results. (A) Mean spectra for tablet types A, B, C &amp; D and background. (B) Mean spectra for tablet types A (teal) and D (pink) for mass range <em>m/z</em> 1065 to <em>m/z</em> 1090. (C-E) Boxplots for peaks assigned as caffeine (C), aspirin (D) and paracetamol (E) for each tablet type and background. (F-L) Transmission Raman spectroscopy results. (F) Mean spectra for tablet type A-D from training data. Peaks for the active ingredients shown in G-L are highlighted with a vertical line. (G-L) Mean spectra and (J-L) boxplots for peaks annotated as active ingredients (G &amp; J) Caffeine, 541 cm<sup>-1</sup>, (H &amp; K) Paracetamol, 847 cm<sup>-1</sup>, and (I &amp; L) Aspirin, 1181 cm<sup>-1</sup>). Boxplots show median (horizontal line), interquartile range (box) and range excluding outliers (whiskers). Points show single scan intensities.</figcaption>
</figure>
</div>
<h3 id="relative-spectral-similarity">Relative spectral similarity</h3>
<p>It is notable that for both DESI MS and Transmission Raman data assessment of active ingredients or film coating, or excipients alone cannot robustly separate all tablet types. We can objectively assess the relative overall spectral similarity between and within tablet types by calculating the cosine distance of each spectrum of the same modality from one another. The cosine similarity matrix for DESI MS (Figure 2A) reveals that the highest cosine similarities are between spectra from tablet type D (D vs. D, 0.97 +/- 0.03) and between spectra from tablet type A (A vs. A, 0.96 +/- 0.02). The low standard deviation of cosine similarity within these tablet types indicates the highly reproducible spectra achieved from these samples. Tablet types C and D have a lower mean cosine similarity and higher variance of similarity within their respective tablet types (B vs. B, 0.77 +/- 0.08. C vs. C 0.85 +/- 0.12). Tablet types B and C are relatively alike, with cosine similarity of 0.71. Conversely, tablet types A and D, are relatively dissimilar to one another (0.45 +/- 0.03)</p>
<p>For Transmission Raman spectroscopy, high cosine similarity is observed within tablet types (all greater than 0.97), However, spectral similarity between tablets A &amp; D is notably higher than for DESI MSI (0.97 +/- 0.01). This is also seen in the high similarity between tables B and C (0.99 +/- 0.01), where the only visible spectral difference is for Caffeine, (541, cm<sup>-1</sup>, present in B, absent in C). All other peaks are visibly identical contributing to the high similarity value.</p>
<p>The visible differences between mean spectra and differences in cosine similarities suggest that this DESI MSI may be amenable for the training of classification algorithms to classify unseen data, but that transmission Raman spectroscopy may be more challenging.</p>
<div id="fig:similarity" class="fignos">
<figure>
<img src="https://github.com/adamjtaylor/tablet-classification/raw/main/content/figures/fig2.png" alt="Figure 2: Relative spectral similarity and classification performance comparision (top row) Cosine similarity matrix for each scan of the training dataset from (A) DESI MS or (B) transmission Raman spectroscopy. (middle row) Cross validation F1 measures for (B) DESI MS and (E) transmission Raman spectroscopy. Bars and labels show mean +/- 1 SD for 10-folds with 10 repeats. (bottom row) Confusion matrix for the test set for (A) DESI MSI classified by a SFM with polynomial kernel or (F) transmission Raman Spectroscopy classified LDA. Colour and labels show proportion of correct classifications." /><figcaption aria-hidden="true"><span>Figure 2:</span> <strong>Relative spectral similarity and classification performance comparision</strong>
(top row) Cosine similarity matrix for each scan of the training dataset from (A) DESI MS or (B) transmission Raman spectroscopy. (middle row) Cross validation F1 measures for (B) DESI MS and (E) transmission Raman spectroscopy. Bars and labels show mean +/- 1 SD for 10-folds with 10 repeats. (bottom row) Confusion matrix for the test set for (A) DESI MSI classified by a SFM with polynomial kernel or (F) transmission Raman Spectroscopy classified LDA. Colour and labels show proportion of correct classifications.</figcaption>
</figure>
</div>
<h3 id="assessment-of-different-classification-algorithms">Assessment of different classification algorithms</h3>
<p>A wide range of classification algorithms are available and have been demonstrated for classification of spectrometric and spectroscopic data. Here we assessed a range of classification algorithms including linear discriminate analysis, tree-based methods, a simple neural network, and support vector machines. A 10-fold cross-validation enables the variance of each algorithm to be assessed. Figure 2B and 2E shows the F1 score of each algorithm for cross-validation for DESI MS and Transmission Raman. The F1 score summarizes the precision and recall of the model with equal weights.</p>
<p>For DESI MSI All tested algorithms performed well with mean F1 scores above 0.88. 4 algorithms provide a mean F1 score above 0.95, these include the two tree-based methods, random forest and boosted tree. However, the two highest performing algorithms are the support vector machines. The use of a SVM with a polynomial kernel provides a F1 score of 0.992 on the cross-validation training set.</p>
<p>For transmission Raman data a linear discriminant analysis (LDA) model is most robust yielding a F1 measure greater than 0.95. Random forest, naïve Bayes and nearest neighbor models also performed well with F1 measures greater than 0.9 a similar level of performance. LDA cross validation was rapid, with the naïve Bayes model having the longest run time (Figure S1b). Unlike for DESI MS, here a SVM with a polynomial kernel performed less well with a F1 measure of 0.876.</p>
<p>The best performing models for each analytical technique (DESI: SVM-poly, Transmission Raman: LDA) were therefore taken forward for further exploration and testing. While not a limitation in a the relatively small datasets demonstrated here it is worth noting the range of training times required for 10-fold cross validation with 10 replicates, ranging from ~100 s for Transmission Raman with a boosted tree to ~500 s for DESI MS with naive Bayes (Figure S1). While a search grid for SVM hyperparameters (degree, cost and scale factor) was assessed, the default parameters proved optimal (Supplementary information B, Fig S3). There are no hyperparameters for the LDA.</p>
<h3 id="test-set-classification-performance">Test set classification performance</h3>
<h4 id="desi-ms-1">DESI MS</h4>
<p>As a support vector machine with a polynomial kernel provided the highest classification performance in the cross-validation of the training set, a model based on this algorithm was trained using the entire training set. This model preserved 410 variables. This model was used to predict tablet type from each scan of an independent test set. The test set was acquired 1 week after the training set was collected, with the instrument in active use and recalibrated in the intervening period. In the test set, tablets not seen in the training set, but from the same type and batch were held under the DESI source for approximately 5 seconds. Scans were individually annotated for known tablet type based on known acquisition order and in reference to the total ion chromatogram (Figure S3a).</p>
<p>The trained SVM model was then used to predict the classification of each scan (Figure S3b). The classification algorithm performed well on the test set with an F1 score of 0.956. A confusion matrix for the ground truth vs. the predicted class (Figure 2C) shows the shows that most misclassifications are between the background and each tablet type. Selected correctly predicted spectra (Figure S3e) show notable characteristics, particularly in the polymer peak envelopes seen for tablet types A and D. A misclassification for one tablet type with another only occurs once (Scan 440, Figure S3f), where a scan labelled as tablet type D is misclassified as tablet type A. The spectrum shows two peaks characteristic of the background and a polymer envelope dissimilar to either tablet type demonstrating that classification performance is weak at the start and end of tablet sampling as the tablet is withdrawn from the source. Highlighting misclassifications on the total ion chromatogram (Figure S3c), shows that most misclassifications occur at the end of a tablet being presented to the mass spectrometer. Selected spectra from scans incorrectly predicted as background (Figure S3f, scans 184 and 380) shows base peaks at <em>m/z</em> 217.11 and 309.10 which are also seen in the mean spectra of the background (Figure S1a), highlighting the importance of accurate ground truth annotation in the assessment of classification models.</p>
<h4 id="transmission-raman">Transmission Raman</h4>
<p>LDA trained on the whole Transmission Raman training set was used to classify an independently acquired test set of transmission Raman data from 24 tablets from the same batch. Classification performance was strong with an F1 score of 0.965 when using LDA (Figure 2F. Here classification is correct for tablet types B and C and A, which are classified correctly despite their high cosine similarity. However, in three cases tablet D misclassified as type A, and in one case type A misclassified as type D. This is somewhat unsurprising as they contain the same active ingredients and amounts, but differ in their film coating, as demonstrated by the notably different polymer profiles seen in the DESI-MS data.</p>
<h3 id="variable-importance">Variable importance</h3>
<h4 id="desi-ms-2">DESI MS</h4>
<p>When assessing the classification performance of algorithms on multidimensional data it is useful to assess the variables that the model has placed importance on. Unlike some classification algorithms such as tree-based methods, support vector machines do not inherently provide a measure of variable importance. We therefore calculated a variance-based variable importance using a feature importance ranking measure (FIRM) approach (2 refs), based on quantifying the relative flatness of each feature. Most variables are seen to have relatively low importance to the model (importance &lt; 0.05), although several variables are prominent in a spectrum of variable importance (Figure 3A). The 30 variables with the highest importance were selected and boxplots of their single scan intensity per tablet type plotted (Figure S5B). Several of these import variables show high intensity for a single class over all others. These include peaks characteristic of the background (<em>m/z</em> 588.42 &amp; <em>m/z</em> 309.20), tablet type B (<em>m/z</em> 693.18, <em>m/z</em> 164.13). Other important variables may be of greater intensity in two types (e/g <em>m/z</em> 821.8 in A and D, <em>m/z</em> 445.04 and B and C), or be present in all but one tablet type (e.g <em>m/z</em> 1013.59).</p>
<p>Considering the importance of each variable and the relative intensity in each class enables and understanding of how the SVM charts a path to accurately divide the classes in multivariate space.</p>
<div id="fig:variable-importance" class="fignos">
<figure>
<img src="https://github.com/adamjtaylor/tablet-classification/raw/main/content/figures/fig3.png" alt="Figure 3: Variable importance for selected classification models (A) FIRM variable importance for SVM-poly classification of DESI MS data. 5 peaks with highest importance are highlighted. (B) Scaling values for each discriminant from LDA of transmission Raman data. Highest absolute loading variables are highlighted with red circles. (B) Boxplots showing LDA scores for the spectra from the transmission Raman training set." /><figcaption aria-hidden="true"><span>Figure 3:</span> <strong>Variable importance for selected classification models</strong>
(A) FIRM variable importance for SVM-poly classification of DESI MS data. 5 peaks with highest importance are highlighted. (B) Scaling values for each discriminant from LDA of transmission Raman data. Highest absolute loading variables are highlighted with red circles. (B) Boxplots showing LDA scores for the spectra from the transmission Raman training set.</figcaption>
</figure>
</div>
<h4 id="transmission-raman-spectroscopy-1">Transmission Raman spectroscopy</h4>
<p>The variable importance of the LDA model was assessed by inspection of the LDA scaling values per wavenumber and component (Figure 3B and S5) and the LDA scores of the training set (Figure 3C). LD1 (proportion of trace: 98.8%) provides wide separation of coated tablet types A and D from uncoated tablets B and C. LD2 (proportion of trace: 0.9 %) broadly separates types C and D from one another. LD3 (proportion of trace: 0.2%) provides some separation between types A and D, although this separation is less distinct.</p>
<p>LDA scaling spectra highlight key features in the transmission Raman spectra that contribute to the classification. Features with a broad range of Raman shifts contribute to each component. Most of the LDA scaling spectral features are consistent with key active ingredient Raman peaks. For example, the highest scoring peak in LD1 occurs at 1159 cm-1 which corresponds to aspirin C-H ring bending<span class="citation" data-cites="BoUXYyf5"><sup><a href="#ref-BoUXYyf5" role="doc-biblioref">52</a></sup></span>. LD1 provides separation of tablet types A and D from B and C, and since A and D contain aspirin and B and C do not, strong weighting on wavenumbers consistent with aspirin Raman peaks would be expected. Several other LDA scaling spectral features are consistent with aspirin peak positions, such as 380, 786 and 1221 cm-1. Other LDA scaling spectral features are consistent with Raman peaks of several ingredients, for example 1555 cm-1 (LD3) which could correspond with spectral features present in all three active ingredients, paracetamol, aspirin, and caffeine, which have peaks centered at 1560, 1557, and 1554 cm-1 respectively<span class="citation" data-cites="1BDd0dQq8 148KCDHV1 BoUXYyf5"><sup><a href="#ref-1BDd0dQq8" role="doc-biblioref">50</a>–<a href="#ref-BoUXYyf5" role="doc-biblioref">52</a></sup></span>.</p>
<h3 id="repeatability-and-reproducibility-of-desi-ms">Repeatability and reproducibility of DESI MS</h3>
<p>In the same acquisition as the test set, tablets of type A (Anadin Extra) from a different manufacturing batch number as that analyzed in the training and test sets were profiled. The SVM model correctly classified 51 of 54 scans annotated as the tablet as being of type A (Supplementary figure S5). The three disagreements were between background and tablet type.</p>
<p>Spectral similarity between acquisitions date and tablet batch was assessed by cosine distance (Figure 5). Cosine distance between both date of acquisition and manufacturing batches were highly similar (Cosine distance &gt;0.9) to the cosine distance of spectra within each data or batch. This suggests that sampling variance is higher than the variance between date of acquisition or manufacturing batch. In this work tablets were held manually, using tweezers. Sample variance may be reduced using a sample mounting system or guide to hold tablets in the same geometry respective to the DESI sprayer during acquisition.</p>
<p>Additional analysis (Supplementary information A) comparing DESI MSI cross-validation classification performance with reduced peak numbers (Figure S2A) and simulated lower mass resolving power (S2B) further demonstrate the robustness of classification performance for DESI MSI, and its potential application on compact mass spectrometers at-line or in the field.</p>
<h2 id="conclusions">Conclusions</h2>
<p>DESI mass spectrometry and transmission Raman spectroscopy are effective methods to acquire characteristic spectral information from solid oral dosage forms containing information about active ingredients and tablet coating components.
These tools provide complementary information: transmission Raman is slightly biased in sensitivity towards the bulk of the tablet, whereas DESI-MS provides sensitive analysis of the surface coating.
A range of machine learning algorithms were found to be capable of classifying tablet type with high precision and recall.
Of these, support vector machines showed the strongest performance for DESI-MSI, while LDA was the most effective for transmission Raman data.</p>
<p>DESI-based classification was primarily based on differences in the tablet coatings, whereas transmission Raman was more sensitive to differences in the active pharmaceutical ingredients due to their higher total content.
Therefore, it may be advantageous to combine these two complementary analytical methods.
Raman spectroscopy’s non-destructive nature makes it potentially more suitable for in-line analysis than DESI-MS, the destructive nature, even if minimally, of which makes it unsuitable for tablets remaining in the supply chain.
Classification performance was retained on datasets of reduced peak number and simulated reduced mass resolving power, indicating the robustness of this approach and its potential applicability to compact mass spectrometers suitable for deployment in counterfeiting, QA/QC, or production line environments.</p>
<h2 id="acknowledgements">Acknowledgements</h2>
<p>The authors thank Ariadna Gonzalez and Caterina Minelli (NPL) for helpful discussion and guidance throughout the project. This work was funded by the UK Government’s Department for Business, Energy, and Industrial Strategy (BEIS), and the National Measurement System as part of NPL’s internal cross-theme project “Innovative Medicines Manufacturing.”</p>
<h2 id="contributions">Contributions</h2>
<p>AT, JB and NB conceived and planned the experiments. AT and AB acquired DESI MS data. NB acquired transmission Raman spectroscopy data. AT performed preprocessing of DESI MS data. DT performed preprocessing of transmission Raman data. AT performed the classification and variable importance analysis. AT, DT, NB and JB interpreted the results. AT, AD, DT, JB and NB wrote the manuscript. All authors discussed the results contributed to the final manuscript.</p>
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references csl-bib-body" role="doc-bibliography">
<div id="ref-Vu1f6wmd" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.xphs.2016.11.011">1S. Laske, A. Paudel, O. Scheibelhofer, S. Sacher, T. Hoermann, J. Khinast, A. Kelly, J. Rantannen, O. Korhonen, F. Stauffer, F. De Leersnyder, T. De Beer, J. Mantanus, P.-F. Chavez, B. Thoorens, P. Ghiotti, M. Schubert, P. Tajarobi, G. Haeffler, S. Lakio, M. Fransson, A. Sparen, S. Abrahmsen-Alami, S. Folestad, A. Funke, I. Backx, B. Kavsek, F. Kjell, M. Michaelis, T. Page, J. Palmer, A. Schaepman, S. Sekulic, S. Hammond, B. Braun and B. Colegrove, <em>Journal of Pharmaceutical Sciences</em>, 2017, <strong>106</strong>, 667–712</a></div>.
</div>
<div id="ref-9XSo86vE" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.ijpharm.2013.01.062">2K. Knop and P. Kleinebudde, <em>International Journal of Pharmaceutics</em>, 2013, <strong>457</strong>, 527–536</a></div>.
</div>
<div id="ref-8hsFY7pn" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1039/c0an00352b">3A. Aina, M. D. Hargreaves, P. Matousek and J. C. Burley, <em>The Analyst</em>, 2010, <strong>135</strong>, 2328</a></div>.
</div>
<div id="ref-1GTjZYdpk" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1007/s00216-006-0950-z">4C. Ricci, L. Nyadong, F. M. Fernandez, P. N. Newton and S. G. Kazarian, <em>Analytical and Bioanalytical Chemistry</em>, 2006, <strong>387</strong>, 551–559</a></div>.
</div>
<div id="ref-1ElQCA6hC" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.talanta.2018.11.027">5E. Deconinck, R. Van Campenhout, C. Aouadi, M. Canfyn, J. L. Bothy, L. Gremeaux, P. Blanckaert and P. Courselle, <em>Talanta</em>, 2019, <strong>195</strong>, 142–151</a></div>.
</div>
<div id="ref-YUItDAnD" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.addr.2015.04.003">6A. Paudel, D. Raijada and J. Rantanen, <em>Advanced Drug Delivery Reviews</em>, 2015, <strong>89</strong>, 3–20</a></div>.
</div>
<div id="ref-8OEKQBYr" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1126/science.1104404">7Z. Takáts, J. M. Wiseman, B. Gologan and R. G. Cooks, <em>Science</em>, 2004, <strong>306</strong>, 471–473</a></div>.
</div>
<div id="ref-nGY6XRTE" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1039/b418697d">8Z. Takáts, I. Cotte-Rodriguez, N. Talaty, H. Chen and R. G. Cooks, <em>Chem. Commun.</em>, 2005, 1950–1952</a></div>.
</div>
<div id="ref-1A0sMbTaa" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1039/c5an00112a">9M. J. Bailey, R. Bradshaw, S. Francese, T. L. Salter, C. Costa, M. Ismail, R. P. Webb, I. Bosman, K. Wolff and M. de Puit, <em>The Analyst</em>, 2015, <strong>140</strong>, 6254–6259</a></div>.
</div>
<div id="ref-tvTiSD66" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1002/jms.2010">10B. Li, N. Bjarnholt, S. H. Hansen and C. Janfelt, <em>Journal of Mass Spectrometry</em>, 2011, <strong>46</strong>, 1241–1246</a></div>.
</div>
<div id="ref-sLghpFT8" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1007/s00216-019-02151-z">11A. Dexter, R. T. Steven, A. Patel, L. A. Dailey, A. J. Taylor, D. Ball, J. Klapwijk, B. Forbes, C. P. Page and J. Bunch, <em>Analytical and Bioanalytical Chemistry</em>, 2019, <strong>411</strong>, 8023–8032</a></div>.
</div>
<div id="ref-1ClzsqfDW" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.bbalip.2011.05.006">12L. S. Eberlin, C. R. Ferreira, A. L. Dill, D. R. Ifa and R. G. Cooks, <em>Biochimica et Biophysica Acta (BBA) - Molecular and Cell Biology of Lipids</em>, 2011, <strong>1811</strong>, 946–960</a></div>.
</div>
<div id="ref-1EVpL1egt" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1021/ac050989d">13H. Chen, N. N. Talaty, Z. Takáts and R. G. Cooks, <em>Analytical Chemistry</em>, 2005, <strong>77</strong>, 6915–6927</a></div>.
</div>
<div id="ref-or5TqRX6" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1002/rcm.2280">14L. A. Leuthold, J.-F. Mandscheff, M. Fathi, C. Giroud, M. Augsburger, E. Varesio and G. Hopfgartner, <em>Rapid Communications in Mass Spectrometry</em>, 2006, <strong>20</strong>, 103–110</a></div>.
</div>
<div id="ref-Cc9lNkvZ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.jasms.2007.11.016">15L. Nyadong, S. Late, M. D. Green, A. Banga and F. M. Fernández, <em>Journal of the American Society for Mass Spectrometry</em>, 2011, <strong>19</strong>, 380–388</a></div>.
</div>
<div id="ref-4iiQhDKH" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1039/b809471c">16L. Nyadong, E. G. Hohenstein, K. Johnson, C. D. Sherrill, M. D. Green and F. M. Fernández, <em>The Analyst</em>, 2008, <strong>133</strong>, 1513</a></div>.
</div>
<div id="ref-1Gg9u8kt2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1039/b517357d">17C. C. Mulligan, N. Talaty and R. G. Cooks, <em>Chemical Communications</em>, 2006, 1709</a></div>.
</div>
<div id="ref-UuPy88nX" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1021/acs.analchem.9b00520">18C. Liu, K. Qi, L. Yao, Y. Xiong, X. Zhang, J. Zang, C. Tian, M. Xu, J. Yang, Z. Lin, Y. Lv, W. Xiong and Y. Pan, <em>Analytical Chemistry</em>, 2019, <strong>91</strong>, 6616–6623</a></div>.
</div>
<div id="ref-TuCEd7AU" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.aca.2014.03.039">19P. S. Gromski, Y. Xu, E. Correa, D. I. Ellis, M. L. Turner and R. Goodacre, <em>Analytica Chimica Acta</em>, 2014, <strong>829</strong>, 1–8</a></div>.
</div>
<div id="ref-e0eRY8v5" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1007/s00521-016-2736-3">20C. Maione, V. C. de O. Souza, L. R. Togni, J. L. da Costa, A. D. Campiglia, F. Barbosa and R. M. Barbosa, <em>Neural Computing and Applications</em>, 2016, <strong>30</strong>, 947–955</a></div>.
</div>
<div id="ref-ydPgvUPy" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.csbj.2020.07.009">21J. Tang, Y. Wang, Y. Luo, J. Fu, Y. Zhang, Y. Li, Z. Xiao, Y. Lou, Y. Qiu and F. Zhu, <em>Computational and Structural Biotechnology Journal</em>, 2020, <strong>18</strong>, 2012–2025</a></div>.
</div>
<div id="ref-qcR4mVsb" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1007/s11306-017-1242-7">22R. Spicer, R. M. Salek, P. Moreno, D. Cañueto and C. Steinbeck, <em>Metabolomics</em>, 2017, <strong>13</strong>, 106</a></div>.
</div>
<div id="ref-O5djQ3nR" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1038/s41416-018-0048-3">23D. L. Phelps, J. Balog, L. F. Gildea, Z. Bodai, A. Savage, M. A. El-Bahrawy, A. V. Speller, F. Rosini, H. Kudo, J. S. McKenzie, R. Brown, Z. Takáts and S. Ghaem-Maghami, <em>British Journal of Cancer</em>, 2018, <strong>118</strong>, 1349–1358</a></div>.
</div>
<div id="ref-g08cPiPx" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1021/acs.jafc.6b01041">24J. Balog, D. Perenyi, C. Guallar-Hoyas, A. Egri, S. D. Pringle, S. Stead, O. P. Chevallier, C. T. Elliott and Z. Takats, <em>Journal of Agricultural and Food Chemistry</em>, 2016, <strong>64</strong>, 4793–4800</a></div>.
</div>
<div id="ref-fKTaMhqM" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1021/ac5046752">25O. Golf, N. Strittmatter, T. Karancsi, S. D. Pringle, A. V. M. Speller, A. Mroz, J. M. Kinross, N. Abbassi-Ghadi, E. A. Jones and Z. Takats, <em>Analytical Chemistry</em>, 2015, <strong>87</strong>, 2527–2534</a></div>.
</div>
<div id="ref-1Dw5WtMxe" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1186/s13058-017-0845-2">26E. R. St John, J. Balog, J. S. McKenzie, M. Rossi, A. Covington, L. Muirhead, Z. Bodai, F. Rosini, A. V. M. Speller, S. Shousha, R. Ramakrishnan, A. Darzi, Z. Takats and D. R. Leff, <em>Breast Cancer Research</em>, 2017, <strong>19</strong>, 59</a></div>.
</div>
<div id="ref-jh1jwVUX" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1039/c3an36843b">27K. Shin and H. Chung, <em>The Analyst</em>, 2013, <strong>138</strong>, 3335</a></div>.
</div>
<div id="ref-13x0WMYJq" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1366/000370206779321463">28P. Matousek and A. W. Parker, <em>Applied Spectroscopy</em>, 2016, <strong>60</strong>, 1353–1357</a></div>.
</div>
<div id="ref-vvUUyBtN" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.jpba.2010.10.029">29K. Buckley and P. Matousek, <em>Journal of Pharmaceutical and Biomedical Analysis</em>, 2011, <strong>55</strong>, 645–652</a></div>.
</div>
<div id="ref-eL1MOkeG" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1366/000370207782597085">30J. Johansson, A. Sparén, O. Svensson, S. Folestad and M. Claybourn, <em>Applied Spectroscopy</em>, 2016, <strong>61</strong>, 1211–1218</a></div>.
</div>
<div id="ref-Qy6BdVBn" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1366/11-06259">31P. Matousek, N. Everall, D. Littlejohn, A. Nordon and M. Bloomfield, <em>Applied Spectroscopy</em>, 2011, <strong>65</strong>, 724–733</a></div>.
</div>
<div id="ref-1EsJaNvp" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.trac.2020.116157">32M. D. Peris-Díaz and A. Krężel, <em>TrAC Trends in Analytical Chemistry</em>, 2021, <strong>135</strong>, 116157</a></div>.
</div>
<div id="ref-ubp6jD46" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1002/jrs.882">33N. Stone, C. Kendall, N. Shepherd, P. Crow and H. Barr, <em>Journal of Raman Spectroscopy</em>, 2002, <strong>33</strong>, 564–573</a></div>.
</div>
<div id="ref-koyIoA0l" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1002/jbio.200810024">34C. Krafft, G. Steiner, C. Beleites and R. Salzer, <em>Journal of Biophotonics</em>, 2009, <strong>2</strong>, 13–28</a></div>.
</div>
<div id="ref-hmGuG8Yj" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1155/2016/1603609">35W. Liu, Z. Sun, J. Chen and C. Jing, <em>Journal of Spectroscopy</em>, 2016, <strong>2016</strong>, 1–6</a></div>.
</div>
<div id="ref-orJZ0nNQ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1002/bip.20448">36K. Gaus, P. Rösch, R. Petry, K.-D. Peschke, O. Ronneberger, H. Burkhardt, K. Baumann and J. Popp, <em>Biopolymers</em>, 2006, <strong>82</strong>, 286–290</a></div>.
</div>
<div id="ref-14piFxIYf" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1002/jrs.5343">37S. Guo, R. Heinke, S. Stöckel, P. Rösch, J. Popp and T. Bocklitz, <em>Journal of Raman Spectroscopy</em>, 2018, <strong>49</strong>, 627–637</a></div>.
</div>
<div id="ref-W0VneN3Y" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1520/jfs15244j">38A. G. Ryder, <em>Journal of Forensic Sciences</em>, 2002, <strong>47</strong>, 15244J</a></div>.
</div>
<div id="ref-6wD6szWs" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.talanta.2010.01.046">39Y. Roggo, K. Degardin and P. Margot, <em>Talanta</em>, 2010, <strong>81</strong>, 988–995</a></div>.
</div>
<div id="ref-d5haWH7e" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.jpba.2006.01.033">40S. Romero-Torres, J. D. Pérez-Ramos, K. R. Morris and E. R. Grant, <em>Journal of Pharmaceutical and Biomedical Analysis</em>, 2006, <strong>41</strong>, 811–819</a></div>.
</div>
<div id="ref-TToYRrJQ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1002/jrs.1621">41M. de Veij, P. Vandenabeele, K. A. Hall, F. M. Fernandez, M. D. Green, N. J. White, A. M. Dondorp, P. N. Newton and L. Moens, <em>Journal of Raman Spectroscopy</em>, 2007, <strong>38</strong>, 181–187</a></div>.
</div>
<div id="ref-16nSdQZQl" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.saa.2019.02.063">42X. Zheng, G. Lv, Y. Zhang, X. Lv, Z. Gao, J. Tang and J. Mo, <em>Spectrochimica Acta Part A: Molecular and Biomolecular Spectroscopy</em>, 2019, <strong>215</strong>, 244–248</a></div>.
</div>
<div id="ref-5Jy2Axu0" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1002/cem.2609">43R. G. Brereton and G. R. Lloyd, <em>Journal of Chemometrics</em>, 2014, <strong>28</strong>, 213–225</a></div>.
</div>
<div id="ref-1FNJpTdZe" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.jpba.2014.01.027">44Q. Gao, Y. Liu, H. Li, H. Chen, Y. Chai and F. Lu, <em>Journal of Pharmaceutical and Biomedical Analysis</em>, 2014, <strong>94</strong>, 58–64</a></div>.
</div>
<div id="ref-BmWj6JSS" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1002/cem.1330">45M. Fransson, J. Johansson, A. Sparén and O. Svensson, <em>Journal of Chemometrics</em>, 2010, <strong>24</strong>, 674–680</a></div>.
</div>
<div id="ref-EWll2pyF" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.21105/joss.01686">46H. Wickham, M. Averick, J. Bryan, W. Chang, L. McGowan, R. François, G. Grolemund, A. Hayes, L. Henry, J. Hester, M. Kuhn, T. Pedersen, E. Miller, S. Bache, K. Müller, J. Ooms, D. Robinson, D. Seidel, V. Spinu, K. Takahashi, D. Vaughan, C. Wilke, K. Woo and H. Yutani, <em>Journal of Open Source Software</em>, 2019, <strong>4</strong>, 1686</a></div>.
</div>
<div id="ref-4zds2yP" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">47 </div><div class="csl-right-inline">Tidymodels, <a href="https://www.tidymodels.org/">https://www.tidymodels.org/</a>, (accessed October 15, 2021).</div>
</div>
<div id="ref-WsAj0wOM" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">48 </div><div class="csl-right-inline"><a href="https://cran.r-project.org/package">https://cran.r-project.org/package</a>.</div>
</div>
<div id="ref-XgRF2vnB" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1021/acs.analchem.6b02139">49A. Dexter, A. M. Race, I. B. Styles and J. Bunch, <em>Analytical Chemistry</em>, 2016, <strong>88</strong>, 10893–10899</a></div>.
</div>
<div id="ref-1BDd0dQq8" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.vibspec.2007.12.016">50M. Baranska and L. M. Proniewicz, <em>Vibrational Spectroscopy</em>, 2008, <strong>48</strong>, 153–157</a></div>.
</div>
<div id="ref-148KCDHV1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.3390/pharmaceutics6040651">51C. Shende, W. Smith, C. Brouillette and S. Farquharson, <em>Pharmaceutics</em>, 2014, <strong>6</strong>, 651–662</a></div>.
</div>
<div id="ref-BoUXYyf5" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin"><a href="https://doi.org/10.1016/j.molstruc.2014.10.079">52E. L. Crowell, Z. A. Dreger and Y. M. Gupta, <em>Journal of Molecular Structure</em>, 2015, <strong>1082</strong>, 29–37</a></div>.
</div>
</div>
<h2 class="page_break_before" id="supplementary-information">Supplementary information</h2>
<h3 id="notes">Notes</h3>
<h4 id="note-a.-reducing-peaks-and-down-binning-data-for-desi-ms">Note A. Reducing peaks and down binning data for DESI MS</h4>
<p>Given the notable spectra differences observed for DESI MS, it is unsurprising that strong classification performance is achieved. In this model comparison we included all detected peaks in the algorithm (although some will be removed automatically as they correlate strongly with other variables. To assess the number of features required to maintain good classification performance we performed 10-fold cross validation using an SVM with a polynomial kernel and default parameters (degree: 1, cost: 0.1, scale factor: 1) on datasets in which fewer peaks were selected (from top 1200 to top 5 peaks). F1 metric was seen to be maintained above 0.97 (Figure S2B), but starts to drop off below 100 peaks, with 5 peaks providing a mean F1 score of 0.61 with a greatly increased variance.</p>
<p>Potential applications of tablet classification may demand the mass spectrometer to be located outside of an analytical chemistry laboratory, either at-line in a manufacturing environment, in an on-site QA/QC lab, on in the field. In these applications the use of a Q-ToF mass spectrometer may not be suitable due to size or cost. Compact, field-deployable mass spectrometers such as ion traps or single quadrupoles may offer reduced mass resolving power. To simulate reduced instrument performance, data were down binned at increasing bin width. No reduction in classification performance is observed when binning at 1 Dalton, simulating single quadrupole mass resolving power (Figure S2B). Classification performance is maintained even at 2 Dalton binning, indicating the applicability of this approach to low mass resolving power data from compact or portable mass spectrometers.</p>
<h4 id="note-b.-model-tuning">Note B. Model tuning</h4>
<p>The hyperparameters of the SVM were tuned on the validation split of the training set over a regular grid with five levels of degree (1 to 5), cost (0.001 to 1) and scale factor (0.001 to 1). All variables and 0.01 Da binning. The F1 measure of each combination is shown in Figure S3. First-degree (linear) and third-degree polynomials with cost and scale factors above 0.001 show high F1 score (all = 1.0) for the validation split. As the default SVM parameters in the kernlab implementation are within this range (degree: 1, cost: 1, scale factor: 1) they were selected to fit a final model using the whole training set.</p>
<p>For Transmission Raman data the best performing classification approach, a linear discriminant analysis, as implemented by the MASS package has no parameters to optimize beyond the prior which was here set to the equal type probability. This would influence the suitability of the LDA for applications in cases such as defect or counterfeit detection where class probability is expected to be unbalanced with an unknown prior.</p>
<h3 class="page_break_before" id="figures">Figures</h3>
<div id="fig:timing" class="fignos">
<figure>
<img src="https://github.com/adamjtaylor/tablet-classification/raw/main/content/figures/fig_timing.png" data-tag="S1" style="width:62.0%" alt="Figure S1: Timings for 10-fold cross validation with 10 replicates for selected classification algorithms for DESI MS (red) and transmission Raman spectroscopy (teal) training data." /><figcaption aria-hidden="true"><span>Figure S1:</span> Timings for 10-fold cross validation with 10 replicates for selected classification algorithms for DESI MS (red) and transmission Raman spectroscopy (teal) training data.</figcaption>
</figure>
</div>
<div id="fig:downbin" class="fignos">
<figure>
<img src="https://github.com/adamjtaylor/tablet-classification/raw/main/content/figures/fig_npeaks_downbin.png" data-tag="S2" style="width:62.0%" alt="Figure S2: Exploration of reduced peak number (A) and down-binning (B) for DESI MS classification. " /><figcaption aria-hidden="true"><span>Figure S2:</span> Exploration of reduced peak number (A) and down-binning (B) for DESI MS classification.<br />
</figcaption>
</figure>
</div>
<div id="fig:tuning" class="fignos">
<figure>
<img src="https://github.com/adamjtaylor/tablet-classification/raw/main/content/figures/fig_tuning.png" data-tag="S3" style="width:75.0%" alt="Figure S3: Model tuning grid for SVM-polynomial kernel for DESI MS data." /><figcaption aria-hidden="true"><span>Figure S3:</span> Model tuning grid for SVM-polynomial kernel for DESI MS data.</figcaption>
</figure>
</div>
<div id="fig:tuning" class="fignos">
<figure>
<img src="https://github.com/adamjtaylor/tablet-classification/raw/main/content/figures/fig_desi_validation_supplementary.png" data-tag="S4" style="width:75.0%" alt="Figure S4: Classification of an independent test set of tablets by DESI MS and an SVM (a) Total ion chromatogram (TIC) plot showing the ground truth labelling per scan for four sampling events per tablet type. Labelling was performed manually in reference to known sampling order and TIC. (b) TIC plot showing tablet type predicted by the SVM for each scan. (c) TIC plot highlighting prediction accuracy per scan (correct: green circles, incorrect: pink triangles). (d) Confusion matrix for each scan of the test set, classified by the SVM. Colour and labels show proportion of correct classifications. (e-f) Spectra for selected scans (labelled in c) that are correctly (e). or incorrectly classified (f)." /><figcaption aria-hidden="true"><span>Figure S4:</span> <strong>Classification of an independent test set of tablets by DESI MS and an SVM</strong><br />
(a) Total ion chromatogram (TIC) plot showing the ground truth labelling per scan for four sampling events per tablet type. Labelling was performed manually in reference to known sampling order and TIC. (b) TIC plot showing tablet type predicted by the SVM for each scan. (c) TIC plot highlighting prediction accuracy per scan (correct: green circles, incorrect: pink triangles). (d) Confusion matrix for each scan of the test set, classified by the SVM. Colour and labels show proportion of correct classifications. (e-f) Spectra for selected scans (labelled in c) that are correctly (e). or incorrectly classified (f).</figcaption>
</figure>
</div>
<div id="fig:replicate_similarity" class="fignos">
<figure>
<img src="https://github.com/adamjtaylor/tablet-classification/raw/main/content/figures/fig_repliacate_similarity.png" data-tag="S5" style="width:50.0%" alt="Figure S5: Cosine similarity matrix for DESI MS sampling of tablet type A from two batches on two days of analysis." /><figcaption aria-hidden="true"><span>Figure S5:</span> Cosine similarity matrix for DESI MS sampling of tablet type A from two batches on two days of analysis.</figcaption>
</figure>
</div>
<div id="fig:desi_vip" class="fignos">
<figure>
<img src="https://github.com/adamjtaylor/tablet-classification/raw/main/content/figures/fig_desi_importance.png" data-tag="S6" style="width:75.0%" alt="Figure S6: Variable importance plot for SVM with polynomial kernel trained on DESI MS data. Peaks with the 30 highest variable importance are highlighted by a black dot). (b) Boxplots for peaks with the 30 highest variable importance values (Line: median, box: Q2 &amp; Q4, whiskers: range excluding outliers, points: single scan intensities)." /><figcaption aria-hidden="true"><span>Figure S6:</span> Variable importance plot for SVM with polynomial kernel trained on DESI MS data. Peaks with the 30 highest variable importance are highlighted by a black dot). (b) Boxplots for peaks with the 30 highest variable importance values (Line: median, box: Q2 &amp; Q4, whiskers: range excluding outliers, points: single scan intensities).</figcaption>
</figure>
</div>
<div id="fig:tr_vip" class="fignos">
<figure>
<img src="https://github.com/adamjtaylor/tablet-classification/raw/main/content/figures/fig_tr_importance.png" data-tag="S7" style="width:75.0%" alt="Figure S7: (a) Spectra of LDA scaling values for each discriminant. Selected variables are highlighted with red circles. Boxplots showing LDA scores for the spectra contained in the training set. (c &amp; d) Boxplots of transmission Raman intensity for the wavenumber bins with the two highest (c) and lowest (d) scaling values per discriminant. Boxplots show mean (bar), first and third quartiles (bar) and range between the lowest and highest values no further than 1.5 times the IQR from the box. Points show individual observations." /><figcaption aria-hidden="true"><span>Figure S7:</span> (a) Spectra of LDA scaling values for each discriminant. Selected variables are highlighted with red circles. Boxplots showing LDA scores for the spectra contained in the training set. (c &amp; d) Boxplots of transmission Raman intensity for the wavenumber bins with the two highest (c) and lowest (d) scaling values per discriminant. Boxplots show mean (bar), first and third quartiles (bar) and range between the lowest and highest values no further than 1.5 times the IQR from the box. Points show individual observations.</figcaption>
</figure>
</div>
<!-- default theme -->

<style>
  /* import google fonts */
  @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
  @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

  /* -------------------------------------------------- */
  /* global */
  /* -------------------------------------------------- */

  /* all elements */
  * {
    /* force sans-serif font unless specified otherwise */
    font-family: "Open Sans", "Helvetica", sans-serif;

    /* prevent text inflation on some mobile browsers */
    -webkit-text-size-adjust: none !important;
    -moz-text-size-adjust: none !important;
    -o-text-size-adjust: none !important;
    text-size-adjust: none !important;
  }

  @media only screen {
    /* "page" element */
    body {
      position: relative;
      box-sizing: border-box;
      font-size: 11pt;
      line-height: 1.5;
      max-width: 8.5in;
      margin: 20px auto;
      padding: 40px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* "page" element */
    body {
      padding: 20px;
      margin: 0;
      border-radius: 0;
      border: none;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
      background: none;
    }
  }

  /* -------------------------------------------------- */
  /* headings */
  /* -------------------------------------------------- */

  /* all headings */
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 20px 0;
    padding: 0;
    font-weight: bold;
  }

  /* biggest heading */
  h1 {
    margin: 40px 0;
    text-align: center;
  }

  /* second biggest heading */
  h2 {
    margin-top: 30px;
    padding-bottom: 5px;
    border-bottom: solid 1px #bdbdbd;
  }

  /* heading font sizes */
  h1 {
    font-size: 2em;
  }
  h2 {
    font-size: 1.5em;
  }
  h3 {
    font-size: 1.35em;
  }
  h4 {
    font-size: 1.25em;
  }
  h5 {
    font-size: 1.15em;
  }
  h6 {
    font-size: 1em;
  }

  /* -------------------------------------------------- */
  /* manuscript header */
  /* -------------------------------------------------- */

  /* manuscript title */
  header > h1 {
    margin: 0;
  }

  /* manuscript title caption text (ie "automatically generated on") */
  header + p {
    text-align: center;
    margin-top: 10px;
  }

  /* -------------------------------------------------- */
  /* text elements */
  /* -------------------------------------------------- */

  /* links */
  a {
    color: #2196f3;
    overflow-wrap: break-word;
  }

  /* superscripts and subscripts */
  sub,
  sup {
    /* prevent from affecting line height */
    line-height: 0;
  }

  /* unordered and ordered lists*/
  ul,
  ol {
    padding-left: 20px;
  }

  /* class for styling text semibold */
  .semibold {
    font-weight: 600;
  }

  /* class for styling elements horizontally left aligned */
  .left {
    display: block;
    text-align: left;
    margin-left: auto;
    margin-right: 0;
    justify-content: left;
  }

  /* class for styling elements horizontally centered */
  .center {
    display: block;
    text-align: center;
    margin-left: auto;
    margin-right: auto;
    justify-content: center;
  }

  /* class for styling elements horizontally right aligned */
  .right {
    display: block;
    text-align: right;
    margin-left: 0;
    margin-right: auto;
    justify-content: right;
  }

  /* -------------------------------------------------- */
  /* section elements */
  /* -------------------------------------------------- */

  /* horizontal divider line */
  hr {
    border: none;
    height: 1px;
    background: #bdbdbd;
  }

  /* paragraphs, horizontal dividers, figures, tables, code */
  p,
  hr,
  figure,
  table,
  pre {
    /* treat all as "paragraphs", with consistent vertical margins */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* figures */
  /* -------------------------------------------------- */

  /* figure */
  figure {
    max-width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure caption */
  figcaption {
    padding: 0;
    padding-top: 10px;
  }

  /* figure image element */
  figure > img,
  figure > svg {
    max-width: 100%;
    display: block;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure auto-number */
  img + figcaption > span:first-of-type,
  svg + figcaption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* tables */
  /* -------------------------------------------------- */

  /* table */
  table {
    border-collapse: collapse;
    border-spacing: 0;
    width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* table cells */
  th,
  td {
    border: solid 1px #bdbdbd;
    padding: 10px;
    /* squash table if too wide for page by forcing line breaks */
    overflow-wrap: break-word;
    word-break: break-word;
  }

  /* header row and even rows */
  th,
  tr:nth-child(2n) {
    background-color: #fafafa;
  }

  /* odd rows */
  tr:nth-child(2n + 1) {
    background-color: #ffffff;
  }

  /* table caption */
  caption {
    text-align: left;
    padding: 0;
    padding-bottom: 10px;
  }

  /* table auto-number */
  table > caption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* code */
  /* -------------------------------------------------- */

  /* multi-line code block */
  pre {
    padding: 10px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
    break-inside: avoid;
    text-align: left;
  }

  /* inline code, ie code within normal text */
  :not(pre) > code {
    padding: 0 4px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
  }

  /* code text */
  /* apply all children, to reach syntax highlighting sub-elements */
  code,
  code * {
    /* force monospace font */
    font-family: "Source Code Pro", "Courier New", monospace;
  }

  /* -------------------------------------------------- */
  /* quotes */
  /* -------------------------------------------------- */

  /* quoted text */
  blockquote {
    margin: 0;
    padding: 0;
    border-left: 4px solid #bdbdbd;
    padding-left: 16px;
    break-inside: avoid;
  }

  /* -------------------------------------------------- */
  /* banners */
  /* -------------------------------------------------- */

  /* info banners */
  .banner {
    box-sizing: border-box;
    display: block;
    position: relative;
    width: 100%;
    margin-top: 20px;
    margin-bottom: 20px;
    padding: 20px;
    text-align: center;
  }

  /* paragraph in banner */
  .banner > p {
    margin: 0;
  }

  /* -------------------------------------------------- */
  /* highlight colors */
  /* -------------------------------------------------- */

  .white {
    background: #ffffff;
  }
  .lightgrey {
    background: #eeeeee;
  }
  .grey {
    background: #757575;
  }
  .darkgrey {
    background: #424242;
  }
  .black {
    background: #000000;
  }
  .lightred {
    background: #ffcdd2;
  }
  .lightyellow {
    background: #ffecb3;
  }
  .lightgreen {
    background: #dcedc8;
  }
  .lightblue {
    background: #e3f2fd;
  }
  .lightpurple {
    background: #f3e5f5;
  }
  .red {
    background: #f44336;
  }
  .orange {
    background: #ff9800;
  }
  .yellow {
    background: #ffeb3b;
  }
  .green {
    background: #4caf50;
  }
  .blue {
    background: #2196f3;
  }
  .purple {
    background: #9c27b0;
  }
  .white,
  .lightgrey,
  .lightred,
  .lightyellow,
  .lightgreen,
  .lightblue,
  .lightpurple,
  .orange,
  .yellow,
  .white a,
  .lightgrey a,
  .lightred a,
  .lightyellow a,
  .lightgreen a,
  .lightblue a,
  .lightpurple a,
  .orange a,
  .yellow a {
    color: #000000;
  }
  .grey,
  .darkgrey,
  .black,
  .red,
  .green,
  .blue,
  .purple,
  .grey a,
  .darkgrey a,
  .black a,
  .red a,
  .green a,
  .blue a,
  .purple a {
    color: #ffffff;
  }

  /* -------------------------------------------------- */
  /* buttons */
  /* -------------------------------------------------- */

  /* class for styling links like buttons */
  .button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    margin: 5px;
    padding: 10px 20px;
    font-size: 0.75em;
    font-weight: 600;
    text-transform: uppercase;
    text-decoration: none;
    letter-spacing: 1px;
    background: none;
    color: #2196f3;
    border: solid 1px #bdbdbd;
    border-radius: 5px;
  }

  /* buttons when hovered */
  .button:hover:not([disabled]),
  .icon_button:hover:not([disabled]) {
    cursor: pointer;
    background: #f5f5f5;
  }

  /* buttons when disabled */
  .button[disabled],
  .icon_button[disabled] {
    opacity: 0.35;
    pointer-events: none;
  }

  /* class for styling buttons containg only single icon */
  .icon_button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    text-decoration: none;
    margin: 0;
    padding: 0;
    background: none;
    border-radius: 5px;
    border: none;
    width: 20px;
    height: 20px;
    min-width: 20px;
    min-height: 20px;
  }

  /* icon button inner svg image */
  .icon_button > svg {
    height: 16px;
  }

  /* -------------------------------------------------- */
  /* icons */
  /* -------------------------------------------------- */

  /* class for styling icons inline with text */
  .inline_icon {
    height: 1em;
    position: relative;
    top: 0.125em;
  }

  /* -------------------------------------------------- */
  /* references */
  /* -------------------------------------------------- */

  .csl-entry {
    margin-top: 15px;
    margin-bottom: 15px;
  }

  /* -------------------------------------------------- */
  /* print control */
  /* -------------------------------------------------- */

  @media print {
    @page {
      /* suggested printing margin */
      margin: 0.5in;
    }

    /* document and "page" elements */
    html,
    body {
      margin: 0;
      padding: 0;
      width: 100%;
      height: 100%;
    }

    /* "page" element */
    body {
      font-size: 11pt !important;
      line-height: 1.35;
    }

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      margin: 15px 0;
    }

    /* figures and tables */
    figure,
    table {
      font-size: 0.85em;
    }

    /* table cells */
    th,
    td {
      padding: 5px;
    }

    /* shrink font awesome icons */
    i.fas,
    i.fab,
    i.far,
    i.fal {
      transform: scale(0.85);
    }

    /* decrease banner margins */
    .banner {
      margin-top: 15px;
      margin-bottom: 15px;
      padding: 15px;
    }

    /* class for centering an element vertically on its own page */
    .page_center {
      margin: auto;
      width: 100%;
      height: 100%;
      display: flex;
      align-items: center;
      vertical-align: middle;
      break-before: page;
      break-after: page;
    }

    /* always insert a page break before the element */
    .page_break_before {
      break-before: page;
    }

    /* always insert a page break after the element */
    .page_break_after {
      break-after: page;
    }

    /* avoid page break before the element */
    .page_break_before_avoid {
      break-before: avoid;
    }

    /* avoid page break after the element */
    .page_break_after_avoid {
      break-after: avoid;
    }

    /* avoid page break inside the element */
    .page_break_inside_avoid {
      break-inside: avoid;
    }
  }

  /* -------------------------------------------------- */
  /* override pandoc css quirks */
  /* -------------------------------------------------- */

  .sourceCode {
    /* prevent unsightly overflow in wide code blocks */
    overflow: auto !important;
  }

  div.sourceCode {
    /* prevent background fill on top-most code block  container */
    background: none !important;
  }

  .sourceCode * {
    /* force consistent line spacing */
    line-height: 1.5 !important;
  }

  div.sourceCode {
    /* style code block margins same as <pre> element */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* tablenos */
  /* -------------------------------------------------- */

  /* tablenos wrapper */
  .tablenos {
    width: 100%;
    margin: 20px 0;
  }

  .tablenos > table {
    /* move margins from table to table_wrapper to allow margin collapsing */
    margin: 0;
  }

  @media only screen {
    /* tablenos wrapper */
    .tablenos {
      /* show scrollbar on tables if necessary to prevent overflow */
      overflow-x: auto !important;
    }

    .tablenos th,
    .tablenos td {
      overflow-wrap: unset !important;
      word-break: unset !important;
    }

    /* table in wrapper */
    .tablenos table,
    .tablenos table * {
      /* don't break table words */
      overflow-wrap: normal !important;
    }
  }
</style>
<!-- 
    Plugin Core

    Functions needed for and shared across all first-party plugins.
-->

<script>
  // get element that is target of hash (from link element or url)
  function getHashTarget(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector(`[id="${id}"]`);
    if (!target) return;

    // if figure or table, modify target to get expected element
    if (id.indexOf("fig:") === 0) target = target.querySelector("figure");
    if (id.indexOf("tbl:") === 0) target = target.querySelector("table");

    return target;
  }

  // get position/dimensions of element or viewport
  function getRectInView(element) {
    let rect = {};
    rect.left = 0;
    rect.top = 0;
    rect.right = document.documentElement.clientWidth;
    rect.bottom = document.documentElement.clientHeight;
    let style = {};

    if (element instanceof HTMLElement) {
      rect = element.getBoundingClientRect();
      style = window.getComputedStyle(element);
    }

    const margin = {};
    margin.left = parseFloat(style.marginLeftWidth) || 0;
    margin.top = parseFloat(style.marginTopWidth) || 0;
    margin.right = parseFloat(style.marginRightWidth) || 0;
    margin.bottom = parseFloat(style.marginBottomWidth) || 0;

    const border = {};
    border.left = parseFloat(style.borderLeftWidth) || 0;
    border.top = parseFloat(style.borderTopWidth) || 0;
    border.right = parseFloat(style.borderRightWidth) || 0;
    border.bottom = parseFloat(style.borderBottomWidth) || 0;

    const newRect = {};
    newRect.left = rect.left + margin.left + border.left;
    newRect.top = rect.top + margin.top + border.top;
    newRect.right = rect.right + margin.right + border.right;
    newRect.bottom = rect.bottom + margin.bottom + border.bottom;
    newRect.width = newRect.right - newRect.left;
    newRect.height = newRect.bottom - newRect.top;

    return newRect;
  }

  // get position of element relative to page
  function getRectInPage(element) {
    const rect = getRectInView(element);
    const body = getRectInView(document.body);

    const newRect = {};
    newRect.left = rect.left - body.left;
    newRect.top = rect.top - body.top;
    newRect.right = rect.right - body.left;
    newRect.bottom = rect.bottom - body.top;
    newRect.width = rect.width;
    newRect.height = rect.height;

    return newRect;
  }

  // get closest element before specified element that matches query
  function firstBefore(element, query) {
    while (element && element !== document.body && !element.matches(query))
      element = element.previousElementSibling || element.parentNode;

    return element;
  }

  // check if element is part of collapsed heading
  function isCollapsed(element) {
    while (element && element !== document.body) {
      if (element.dataset.collapsed === "true") return true;
      element = element.parentNode;
    }
    return false;
  }

  // expand any collapsed parent containers of element if necessary
  function expandElement(element) {
    if (isCollapsed(element)) {
      // accordion plugin
      const heading = firstBefore(element, "h2");
      if (heading) heading.click();
      // details/summary HTML element
      const summary = firstBefore(element, "summary");
      if (summary) summary.click();
    }
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);

    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // get list of elements after a start element up to element matching query
  function nextUntil(element, query, exclude) {
    const elements = [];
    while (((element = element.nextElementSibling), element)) {
      if (element.matches(query)) break;
      if (!element.matches(exclude)) elements.push(element);
    }
    return elements;
  }
</script>
<!--
  Accordion Plugin

  Allows sections of content under h2 headings to be collapsible.
-->

<script type="module">
  // whether to always start expanded ('false'), always start collapsed
  // ('true'), or start collapsed when screen small ('auto')
  const startCollapsed = "auto";

  // start script
  function start() {
    // run through each <h2> heading
    const headings = document.querySelectorAll("h2");
    for (const heading of headings) {
      addArrow(heading);

      // start expanded/collapsed based on option
      if (
        startCollapsed === "true" ||
        (startCollapsed === "auto" && isSmallScreen()) ||
        heading.dataset.collapsed === "true"
      )
        collapseHeading(heading);
      else expandElement(heading);
    }

    // attach hash change listener to window
    window.addEventListener("hashchange", onHashChange);
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) goToElement(target);
  }

  // add arrow to heading
  function addArrow(heading) {
    // add arrow button
    const arrow = document.createElement("button");
    arrow.innerHTML = document.querySelector(".icon_angle_down").innerHTML;
    arrow.classList.add("icon_button", "accordion_arrow");
    heading.insertBefore(arrow, heading.firstChild);

    // attach click listener to heading and button
    heading.addEventListener("click", onHeadingClick);
    arrow.addEventListener("click", onArrowClick);
  }

  // determine if on mobile-like device with small screen
  function isSmallScreen() {
    return Math.min(window.innerWidth, window.innerHeight) < 480;
  }

  // when <h2> heading is clicked
  function onHeadingClick(event) {
    // only collapse if <h2> itself is target of click (eg, user did
    // not click on anchor within <h2>)
    if (event.target === this) toggleCollapse(this);
  }

  // when arrow button is clicked
  function onArrowClick() {
    toggleCollapse(this.parentNode);
  }

  // collapse section if expanded, expand if collapsed
  function toggleCollapse(heading) {
    if (heading.dataset.collapsed === "false") collapseHeading(heading);
    else expandElement(heading);
  }

  // elements to exclude from collapse, such as table of contents panel,
  // hypothesis panel, etc
  const exclude = "#toc_panel, div.annotator-frame, #lightbox_overlay";

  // collapse section
  function collapseHeading(heading) {
    heading.setAttribute("data-collapsed", "true");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "true");
  }

  // expand section
  function expandElement(heading) {
    heading.setAttribute("data-collapsed", "false");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "false");
  }

  // get list of elements between this <h2> and next <h2> or <h1>
  // ("children" of the <h2> section)
  function getChildren(heading) {
    return nextUntil(heading, "h2, h1", exclude);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
  <!-- modified from: https://fontawesome.com/icons/angle-down -->
  <svg width="16" height="16" viewBox="0 0 448 512">
    <path
      fill="currentColor"
      d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* accordion arrow button */
    .accordion_arrow {
      margin-right: 10px;
    }

    /* arrow icon when <h2> data-collapsed attribute true */
    h2[data-collapsed="true"] > .accordion_arrow > svg {
      transform: rotate(-90deg);
    }

    /* all elements (except <h2>'s) when data-collapsed attribute true */
    *:not(h2)[data-collapsed="true"] {
      display: none;
    }

    /* accordion arrow button when hovered and <h2>'s when hovered */
    .accordion_arrow:hover,
    h2[data-collapsed="true"]:hover,
    h2[data-collapsed="false"]:hover {
      cursor: pointer;
    }
  }

  /* always hide accordion arrow button on print */
  @media only print {
    .accordion_arrow {
      display: none;
    }
  }
</style>
<!--
  Anchors Plugin

  Adds an anchor next to each of a certain type of element that provides a
  human-readable url to that specific item/position in the document (e.g.
  "manuscript.html#abstract"). It also makes it such that scrolling out of view
  of a target removes its identifier from the url.
-->

<script type="module">
  // which types of elements to add anchors next to, in "document.querySelector"
  // format
  const typesQuery =
    'h1, h2, h3, div[id^="fig:"], div[id^="tbl:"], span[id^="eq:"]';

  // start script
  function start() {
    // add anchor to each element of specified types
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) addAnchor(element);

    // attach scroll listener to window
    window.addEventListener("scroll", onScroll);
  }

  // when window is scrolled
  function onScroll() {
    // if url has hash and user has scrolled out of view of hash
    // target, remove hash from url
    const tolerance = 100;
    const target = getHashTarget();
    if (target) {
      if (
        target.getBoundingClientRect().top > window.innerHeight + tolerance ||
        target.getBoundingClientRect().bottom < 0 - tolerance
      )
        history.pushState(null, null, " ");
    }
  }

  // add anchor to element
  function addAnchor(element) {
    let addTo; // element to add anchor button to

    // if figure or table, modify withId and addTo to get expected
    // elements
    if (element.id.indexOf("fig:") === 0) {
      addTo = element.querySelector("figcaption");
    } else if (element.id.indexOf("tbl:") === 0) {
      addTo = element.querySelector("caption");
    } else if (element.id.indexOf("eq:") === 0) {
      addTo = element.querySelector(".eqnos-number");
    }

    addTo = addTo || element;
    const id = element.id || null;

    // do not add anchor if element doesn't have assigned id.
    // id is generated by pandoc and is assumed to be unique and
    // human-readable
    if (!id) return;

    // create anchor button
    const anchor = document.createElement("a");
    anchor.innerHTML = document.querySelector(".icon_link").innerHTML;
    anchor.title = "Link to this part of the document";
    anchor.classList.add("icon_button", "anchor");
    anchor.dataset.ignore = "true";
    anchor.href = "#" + id;
    addTo.appendChild(anchor);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- link icon -->

<template class="icon_link">
  <!-- modified from: https://fontawesome.com/icons/link -->
  <svg width="16" height="16" viewBox="0 0 512 512">
    <path
      fill="currentColor"
      d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* anchor button */
    .anchor {
      opacity: 0;
      margin-left: 5px;
    }

    /* anchor buttons within <h2>'s */
    h2 .anchor {
      margin-left: 10px;
    }

    /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
    *:hover > .anchor,
    .anchor:hover,
    .anchor:focus {
      opacity: 1;
    }

    /* anchor button when hovered */
    .anchor:hover {
      cursor: pointer;
    }
  }

  /* always show anchor button on devices with no mouse/hover ability */
  @media (hover: none) {
    .anchor {
      opacity: 1;
    }
  }

  /* always hide anchor button on print */
  @media only print {
    .anchor {
      display: none;
    }
  }
</style>
<!-- 
    Attributes Plugin

    Allows arbitrary HTML attributes to be attached to (almost) any element.
    Place an HTML comment inside or next to the desired element with the content:
    $attribute="value"
-->

<script type="module">
  // start script
  function start() {
    // get list of comments in document
    const comments = findComments();

    for (const comment of comments)
      if (comment.parentElement)
        addAttributes(comment.parentElement, comment.nodeValue.trim());
  }

  // add html attributes to specified element based on string of
  // html attributes and values
  function addAttributes(element, text) {
    // regex's for finding attribute/value pairs in the format of
    // attribute="value" or attribute='value
    const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
    const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

    // loop through attribute/value pairs
    let match;
    while ((match = text.match(regex2) || text.match(regex1))) {
      // get attribute and value from regex capture groups
      let attribute = match[1];
      let value = match[2];

      // remove from string
      text = text.substring(match.index + match[0].length);

      if (!attribute || !value) break;

      // set attribute of parent element
      try {
        element.setAttribute(attribute, value);
      } catch (error) {
        console.log(error);
      }

      // special case for colspan
      if (attribute === "colspan") removeTableCells(element, value);
    }
  }

  // get list of comment elements in document
  function findComments() {
    const comments = [];

    // iterate over comment nodes in document
    function acceptNode(node) {
      return NodeFilter.FILTER_ACCEPT;
    }
    const iterator = document.createNodeIterator(
      document.body,
      NodeFilter.SHOW_COMMENT,
      acceptNode
    );
    let node;
    while ((node = iterator.nextNode())) comments.push(node);

    return comments;
  }

  // remove certain number of cells after specified cell
  function removeTableCells(cell, number) {
    number = parseInt(number);
    if (!number) return;

    // remove elements
    for (; number > 1; number--) {
      if (cell.nextElementSibling) cell.nextElementSibling.remove();
    }
  }

  // start script on DOMContentLoaded instead of load to ensure this plugins
  // runs before other plugins
  window.addEventListener("DOMContentLoaded", start);
</script>
<!--
  Jump to First Plugin

  Adds a button next to each reference entry, figure, and table that jumps the
  page to the first occurrence of a link to that item in the manuscript.
-->

<script type="module">
  // whether to add buttons next to reference entries
  const references = "true";
  // whether to add buttons next to figures
  const figures = "true";
  // whether to add buttons next to tables
  const tables = "true";

  // start script
  function start() {
    if (references !== "false")
      makeButtons(`div[id^="ref-"]`, ".csl-left-margin", "reference");
    if (figures !== "false")
      makeButtons(`div[id^="fig:"]`, "figcaption", "figure");
    if (tables !== "false") makeButtons(`div[id^="tbl:"]`, "caption", "table");
  }

  // when jump button clicked
  function onButtonClick() {
    const first = getFirstOccurrence(this.dataset.id);
    if (!first) return;

    // update url hash so navigating "back" in history will return user to button
    window.location.hash = this.dataset.id;
    // scroll to link
    const timeout = function () {
      goToElement(first, window.innerHeight * 0.5);
    };
    window.setTimeout(timeout, 0);
  }

  // get first occurrence of link to item in document
  function getFirstOccurrence(id) {
    let query = "a";
    query += '[href="#' + id + '"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelector(query);
  }

  // add button next to each reference entry, figure, or table
  function makeButtons(query, containerQuery, subject) {
    const elements = document.querySelectorAll(query);
    for (const element of elements) {
      const id = element.id;
      const buttonContainer = element.querySelector(containerQuery);
      const first = getFirstOccurrence(id);

      // if can't find link to reference or place to put button, ignore
      if (!first || !buttonContainer) continue;

      // make jump button
      let button = document.createElement("button");
      button.classList.add("icon_button", "jump_arrow");
      button.title = `Jump to the first occurrence of this ${subject} in the document`;
      const icon = document.querySelector(".icon_angle_double_up");
      button.innerHTML = icon.innerHTML;
      button.dataset.id = id;
      button.dataset.ignore = "true";
      button.addEventListener("click", onButtonClick);
      buttonContainer.prepend(button);
    }
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
  <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
  <svg width="16" height="16" viewBox="0 0 320 512">
    <path
      fill="currentColor"
      d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* jump button */
    .jump_arrow {
      position: relative;
      top: 0.125em;
      margin-right: 5px;
    }
  }

  /* always hide jump button on print */
  @media only print {
    .jump_arrow {
      display: none;
    }
  }
</style>
<!-- 
    Lightbox Plugin

    Makes it such that when a user clicks on an image, the image fills the
    screen and the user can pan/drag/zoom the image and navigate between other
    images in the document.
-->

<script type="module">
  // list of possible zoom/scale factors
  const zooms =
    "0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1, 1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8";
  // whether to fit image to view ('fit'), display at 100% and shrink if
  // necessary ('shrink'), or always display at 100% ('100')
  const defaultZoom = "fit";
  // whether to zoom in/out toward center of view ('true') or mouse ('false')
  const centerZoom = "false";

  // start script
  function start() {
    // run through each <img> element
    const imgs = document.querySelectorAll("figure > img");
    let count = 1;
    for (const img of imgs) {
      img.classList.add("lightbox_document_img");
      img.dataset.number = count;
      img.dataset.total = imgs.length;
      img.addEventListener("click", openLightbox);
      count++;
    }

    // attach mouse and key listeners to window
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("keyup", onKeyUp);
  }

  // when mouse is moved anywhere in window
  function onWindowMouseMove(event) {
    window.mouseX = event.clientX;
    window.mouseY = event.clientY;
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("lightbox_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("lightbox_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeLightbox();
        break;
    }
  }

  // open lightbox
  function openLightbox() {
    const lightbox = makeLightbox(this);
    if (!lightbox) return;

    blurBody(lightbox);
    document.body.appendChild(lightbox);
  }

  // make lightbox
  function makeLightbox(img) {
    // delete lightbox if it exists, start fresh
    closeLightbox();

    // create screen overlay containing lightbox
    const overlay = document.createElement("div");
    overlay.id = "lightbox_overlay";

    // create image info boxes
    const numberInfo = document.createElement("div");
    const zoomInfo = document.createElement("div");
    numberInfo.id = "lightbox_number_info";
    zoomInfo.id = "lightbox_zoom_info";

    // create container for image
    const imageContainer = document.createElement("div");
    imageContainer.id = "lightbox_image_container";
    const lightboxImg = makeLightboxImg(
      img,
      imageContainer,
      numberInfo,
      zoomInfo
    );
    imageContainer.appendChild(lightboxImg);

    // create bottom container for caption and navigation buttons
    const bottomContainer = document.createElement("div");
    bottomContainer.id = "lightbox_bottom_container";
    const caption = makeCaption(img);
    const prevButton = makePrevButton(img);
    const nextButton = makeNextButton(img);
    bottomContainer.appendChild(prevButton);
    bottomContainer.appendChild(caption);
    bottomContainer.appendChild(nextButton);

    // attach top middle and bottom to overlay
    overlay.appendChild(numberInfo);
    overlay.appendChild(zoomInfo);
    overlay.appendChild(imageContainer);
    overlay.appendChild(bottomContainer);

    return overlay;
  }

  // make <img> object that is intuitively draggable and zoomable
  function makeLightboxImg(sourceImg, container, numberInfoBox, zoomInfoBox) {
    // create copy of source <img>
    const img = sourceImg.cloneNode(true);
    img.classList.remove("lightbox_document_img");
    img.removeAttribute("id");
    img.removeAttribute("width");
    img.removeAttribute("height");
    img.style.position = "unset";
    img.style.margin = "0";
    img.style.padding = "0";
    img.style.width = "";
    img.style.height = "";
    img.style.minWidth = "";
    img.style.minHeight = "";
    img.style.maxWidth = "";
    img.style.maxHeight = "";
    img.id = "lightbox_img";

    // build sorted list of zoomSteps
    const zoomSteps = zooms.split(/[^0-9.]/).map((step) => parseFloat(step));
    zoomSteps.sort((a, b) => a - b);

    // <img> object property variables
    let zoom = 1;
    let translateX = 0;
    let translateY = 0;
    let clickMouseX = undefined;
    let clickMouseY = undefined;
    let clickTranslateX = undefined;
    let clickTranslateY = undefined;

    updateNumberInfo();

    // update image numbers displayed in info box
    function updateNumberInfo() {
      numberInfoBox.innerHTML =
        sourceImg.dataset.number + " of " + sourceImg.dataset.total;
    }

    // update zoom displayed in info box
    function updateZoomInfo() {
      let zoomInfo = zoom * 100;
      if (!Number.isInteger(zoomInfo)) zoomInfo = zoomInfo.toFixed(2);
      zoomInfoBox.innerHTML = zoomInfo + "%";
    }

    // move to closest zoom step above current zoom
    const zoomIn = function () {
      for (const zoomStep of zoomSteps) {
        if (zoomStep > zoom) {
          zoom = zoomStep;
          break;
        }
      }
      updateTransform();
    };

    // move to closest zoom step above current zoom
    const zoomOut = function () {
      zoomSteps.reverse();
      for (const zoomStep of zoomSteps) {
        if (zoomStep < zoom) {
          zoom = zoomStep;
          break;
        }
      }
      zoomSteps.reverse();

      updateTransform();
    };

    // update display of <img> based on scale/translate properties
    const updateTransform = function () {
      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      // get new width/height after scale
      const rect = img.getBoundingClientRect();
      // limit translate
      translateX = Math.max(translateX, -rect.width / 2);
      translateX = Math.min(translateX, rect.width / 2);
      translateY = Math.max(translateY, -rect.height / 2);
      translateY = Math.min(translateY, rect.height / 2);

      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      updateZoomInfo();
    };

    // fit <img> to container
    const fit = function () {
      // no x/y offset, 100% zoom by default
      translateX = 0;
      translateY = 0;
      zoom = 1;

      // widths of <img> and container
      const imgWidth = img.naturalWidth;
      const imgHeight = img.naturalHeight;
      const containerWidth = parseFloat(
        window.getComputedStyle(container).width
      );
      const containerHeight = parseFloat(
        window.getComputedStyle(container).height
      );

      // how much zooming is needed to fit <img> to container
      const xRatio = imgWidth / containerWidth;
      const yRatio = imgHeight / containerHeight;
      const maxRatio = Math.max(xRatio, yRatio);
      const newZoom = 1 / maxRatio;

      // fit <img> to container according to option
      if (defaultZoom === "shrink") {
        if (maxRatio > 1) zoom = newZoom;
      } else if (defaultZoom === "fit") zoom = newZoom;

      updateTransform();
    };

    // when mouse wheel is rolled anywhere in container
    const onContainerWheel = function (event) {
      if (!event) return;

      // let ctrl + mouse wheel to zoom behave as normal
      if (event.ctrlKey) return;

      // prevent normal scroll behavior
      event.preventDefault();
      event.stopPropagation();

      // point around which to scale img
      const viewRect = container.getBoundingClientRect();
      const viewX = (viewRect.left + viewRect.right) / 2;
      const viewY = (viewRect.top + viewRect.bottom) / 2;
      const originX = centerZoom === "true" ? viewX : mouseX;
      const originY = centerZoom === "true" ? viewY : mouseY;

      // get point on image under origin
      const oldRect = img.getBoundingClientRect();
      const oldPercentX = (originX - oldRect.left) / oldRect.width;
      const oldPercentY = (originY - oldRect.top) / oldRect.height;

      // increment/decrement zoom
      if (event.deltaY < 0) zoomIn();
      if (event.deltaY > 0) zoomOut();

      // get offset between previous image point and origin
      const newRect = img.getBoundingClientRect();
      const offsetX = originX - (newRect.left + newRect.width * oldPercentX);
      const offsetY = originY - (newRect.top + newRect.height * oldPercentY);

      // translate image to keep image point under origin
      translateX += offsetX;
      translateY += offsetY;

      // perform translate
      updateTransform();
    };

    // when container is clicked
    function onContainerClick(event) {
      // if container itself is target of click, and not other
      // element above it
      if (event.target === this) closeLightbox();
    }

    // when mouse button is pressed on image
    const onImageMouseDown = function (event) {
      // store original mouse position relative to image
      clickMouseX = window.mouseX;
      clickMouseY = window.mouseY;
      clickTranslateX = translateX;
      clickTranslateY = translateY;
      event.stopPropagation();
      event.preventDefault();
    };

    // when mouse button is released anywhere in window
    const onWindowMouseUp = function (event) {
      // reset original mouse position
      clickMouseX = undefined;
      clickMouseY = undefined;
      clickTranslateX = undefined;
      clickTranslateY = undefined;

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mouseup", onWindowMouseUp);
    };

    // when mouse is moved anywhere in window
    const onWindowMouseMove = function (event) {
      if (
        clickMouseX === undefined ||
        clickMouseY === undefined ||
        clickTranslateX === undefined ||
        clickTranslateY === undefined
      )
        return;

      // offset image based on original and current mouse position
      translateX = clickTranslateX + window.mouseX - clickMouseX;
      translateY = clickTranslateY + window.mouseY - clickMouseY;
      updateTransform();
      event.preventDefault();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mousemove", onWindowMouseMove);
    };

    // when window is resized
    const onWindowResize = function (event) {
      fit();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("resize", onWindowResize);
    };

    // attach the necessary event listeners
    img.addEventListener("dblclick", fit);
    img.addEventListener("mousedown", onImageMouseDown);
    container.addEventListener("wheel", onContainerWheel);
    container.addEventListener("mousedown", onContainerClick);
    container.addEventListener("touchstart", onContainerClick);
    window.addEventListener("mouseup", onWindowMouseUp);
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("resize", onWindowResize);

    // run fit() after lightbox atttached to document and <img> Loaded
    // so needed container and img dimensions available
    img.addEventListener("load", fit);

    return img;
  }

  // make caption
  function makeCaption(img) {
    const caption = document.createElement("div");
    caption.id = "lightbox_caption";
    const captionSource = img.nextElementSibling;
    if (captionSource.tagName.toLowerCase() === "figcaption") {
      const captionCopy = makeCopy(captionSource);
      caption.innerHTML = captionCopy.innerHTML;
    }

    caption.addEventListener("touchstart", function (event) {
      event.stopPropagation();
    });

    return caption;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // make button to jump to previous image in document
  function makePrevButton(img) {
    const prevButton = document.createElement("button");
    prevButton.id = "lightbox_prev_button";
    prevButton.title = "Jump to the previous image in the document [←]";
    prevButton.classList.add("icon_button", "lightbox_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;

    // attach click listeners to button
    prevButton.addEventListener("click", function () {
      getPrevImg(img).click();
    });

    return prevButton;
  }

  // make button to jump to next image in document
  function makeNextButton(img) {
    const nextButton = document.createElement("button");
    nextButton.id = "lightbox_next_button";
    nextButton.title = "Jump to the next image in the document [→]";
    nextButton.classList.add("icon_button", "lightbox_button");
    nextButton.innerHTML = document.querySelector(
      ".icon_caret_right"
    ).innerHTML;

    // attach click listeners to button
    nextButton.addEventListener("click", function () {
      getNextImg(img).click();
    });

    return nextButton;
  }

  // get previous image in document
  function getPrevImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if < 1
    if (index - 1 >= 0) index--;
    else index = imgs.length - 1;
    return imgs[index];
  }

  // get next image in document
  function getNextImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if > total
    if (index + 1 <= imgs.length - 1) index++;
    else index = 0;
    return imgs[index];
  }

  // close lightbox
  function closeLightbox() {
    focusBody();

    const lightbox = document.getElementById("lightbox_overlay");
    if (lightbox) lightbox.remove();
  }

  // make all elements behind lightbox non-focusable
  function blurBody(overlay) {
    const all = document.querySelectorAll("*");
    for (const element of all) element.tabIndex = -1;
    document.body.classList.add("body_no_scroll");
  }

  // make all elements focusable again
  function focusBody() {
    const all = document.querySelectorAll("*");
    for (const element of all) element.removeAttribute("tabIndex");
    document.body.classList.remove("body_no_scroll");
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* regular <img> in document when hovered */
    img.lightbox_document_img:hover {
      cursor: pointer;
    }

    .body_no_scroll {
      overflow: hidden !important;
    }

    /* screen overlay */
    #lightbox_overlay {
      display: flex;
      flex-direction: column;
      position: fixed;
      left: 0;
      top: 0;
      right: 0;
      bottom: 0;
      background: rgba(0, 0, 0, 0.75);
      z-index: 3;
    }

    /* middle area containing lightbox image */
    #lightbox_image_container {
      flex-grow: 1;
      display: flex;
      justify-content: center;
      align-items: center;
      overflow: hidden;
      position: relative;
      padding: 20px;
    }

    /* bottom area containing caption */
    #lightbox_bottom_container {
      display: flex;
      justify-content: center;
      align-items: center;
      height: 100px;
      min-height: 100px;
      max-height: 100px;
      background: rgba(0, 0, 0, 0.5);
    }

    /* image number info text box */
    #lightbox_number_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      left: 2px;
      top: 0;
      z-index: 4;
    }

    /* zoom info text box */
    #lightbox_zoom_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      right: 2px;
      top: 0;
      z-index: 4;
    }

    /* copy of image caption */
    #lightbox_caption {
      box-sizing: border-box;
      display: inline-block;
      width: 100%;
      max-height: 100%;
      padding: 10px 0;
      text-align: center;
      overflow-y: auto;
      color: #ffffff;
    }

    /* navigation previous/next button */
    .lightbox_button {
      width: 100px;
      height: 100%;
      min-width: 100px;
      min-height: 100%;
      color: #ffffff;
    }

    /* navigation previous/next button when hovered */
    .lightbox_button:hover {
      background: none !important;
    }

    /* navigation button icon */
    .lightbox_button > svg {
      height: 25px;
    }

    /* figure auto-number */
    #lightbox_caption > span:first-of-type {
      font-weight: bold;
      margin-right: 5px;
    }

    /* lightbox image when hovered */
    #lightbox_img:hover {
      cursor: grab;
    }

    /* lightbox image when grabbed */
    #lightbox_img:active {
      cursor: grabbing;
    }
  }

  /* when on screen < 480px wide */
  @media only screen and (max-width: 480px) {
    /* make navigation buttons skinnier on small screens to make more room for caption text */
    .lightbox_button {
      width: 50px;
      min-width: 50px;
    }
  }

  /* always hide lightbox on print */
  @media only print {
    #lightbox_overlay {
      display: none;
    }
  }
</style>
<!-- 
  Link Highlight Plugin

  Makes it such that when a user hovers or focuses a link, other links that have
  the same target will be highlighted. It also makes it such that when clicking
  a link, the target of the link (eg reference, figure, table) is briefly
  highlighted.
-->

<script type="module">
  // whether to also highlight links that go to external urls
  const externalLinks = "false";
  // whether user must click off to unhighlight instead of just
  // un-hovering
  const clickUnhighlight = "false";
  // whether to also highlight links that are unique
  const highlightUnique = "true";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach mouse and focus listeners to link
      link.addEventListener("mouseenter", onLinkFocus);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("mouseleave", onLinkUnhover);
    }

    // attach click and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("hashchange", onHashChange);

    // run hash change on window load in case user has navigated
    // directly to hash
    onHashChange();
  }

  // when link is focused (tabbed to) or hovered
  function onLinkFocus() {
    highlight(this);
  }

  // when link is unhovered
  function onLinkUnhover() {
    if (clickUnhighlight !== "true") unhighlightAll();
  }

  // when the mouse is clicked anywhere in window
  function onClick(event) {
    unhighlightAll();
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) glowElement(target);
  }

  // start glow sequence on an element
  function glowElement(element) {
    const startGlow = function () {
      onGlowEnd();
      element.dataset.glow = "true";
      element.addEventListener("animationend", onGlowEnd);
    };
    const onGlowEnd = function () {
      element.removeAttribute("data-glow");
      element.removeEventListener("animationend", onGlowEnd);
    };
    startGlow();
  }

  // highlight link and all others with same target
  function highlight(link) {
    // force unhighlight all to start fresh
    unhighlightAll();

    // get links with same target
    if (!link) return;
    const sameLinks = getSameLinks(link);

    // if link unique and option is off, exit and don't highlight
    if (sameLinks.length <= 1 && highlightUnique !== "true") return;

    // highlight all same links, and "select" (special highlight) this
    // one
    for (const sameLink of sameLinks) {
      if (sameLink === link) sameLink.setAttribute("data-selected", "true");
      else sameLink.setAttribute("data-highlighted", "true");
    }
  }

  // unhighlight all links
  function unhighlightAll() {
    const links = getLinks();
    for (const link of links) {
      link.setAttribute("data-selected", "false");
      link.setAttribute("data-highlighted", "false");
    }
  }

  // get links with same target
  function getSameLinks(link) {
    const results = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        results.push(otherLink);
    }
    return results;
  }

  // get all links of types we wish to handle
  function getLinks() {
    let query = "a";
    if (externalLinks !== "true") query += '[href^="#"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelectorAll(query);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<style>
  @media only screen {
    /* anything with data-highlighted attribute true */
    [data-highlighted="true"] {
      background: #ffeb3b;
    }

    /* anything with data-selected attribute true */
    [data-selected="true"] {
      background: #ff8a65 !important;
    }

    /* animation definition for glow */
    @keyframes highlight_glow {
      0% {
        background: none;
      }
      10% {
        background: #bbdefb;
      }
      100% {
        background: none;
      }
    }

    /* anything with data-glow attribute true */
    [data-glow="true"] {
      animation: highlight_glow 2s;
    }
  }
</style>
<!--
  Table of Contents Plugin

  Provides a "table of contents" (toc) panel on the side of the document that
  allows the user to conveniently navigate between sections of the document.
-->

<script type="module">
  // which types of elements to add links for, in "document.querySelector" format
  const typesQuery = "h1, h2, h3";
  // whether toc starts open. use 'true' or 'false', or 'auto' to
  // use 'true' behavior when screen wide enough and 'false' when not
  const startOpen = "false";
  // whether toc closes when clicking on toc link. use 'true' or
  // 'false', or 'auto' to use 'false' behavior when screen wide
  // enough and 'true' when not
  const clickClose = "auto";
  // if list item is more than this many characters, text will be
  // truncated
  const charLimit = "50";
  // whether or not to show bullets next to each toc item
  const bullets = "false";

  // start script
  function start() {
    // make toc panel and populate with entries (links to document
    // sections)
    const panel = makePanel();
    if (!panel) return;
    makeEntries(panel);
    // attach panel to document after making entries, so 'toc' heading
    // in panel isn't included in toc
    document.body.insertBefore(panel, document.body.firstChild);

    // initial panel state
    if (startOpen === "true" || (startOpen === "auto" && !isSmallScreen()))
      openPanel();
    else closePanel();

    // attach click, scroll, and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("scroll", onScroll);
    window.addEventListener("hashchange", onScroll);
    window.addEventListener("keyup", onKeyUp);
    onScroll();

    // add class to push document body down out of way of toc button
    document.body.classList.add("toc_body_nudge");
  }

  // determine if screen wide enough to fit toc panel
  function isSmallScreen() {
    // in default theme:
    // 816px = 8.5in = width of "page" (<body>) element
    // 260px = min width of toc panel (*2 for both sides of <body>)
    return window.innerWidth < 816 + 260 * 2;
  }

  // when mouse is clicked anywhere in window
  function onClick() {
    if (isSmallScreen()) closePanel();
  }

  // when window is scrolled or hash changed
  function onScroll() {
    highlightViewed();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    // close on esc
    if (event.key === "Escape") closePanel();
  }

  // find entry of currently viewed document section in toc and highlight
  function highlightViewed() {
    const firstId = getFirstInView(typesQuery);

    // get toc entries (links), unhighlight all, then highlight viewed
    const list = document.getElementById("toc_list");
    if (!firstId || !list) return;
    const links = list.querySelectorAll("a");
    for (const link of links) link.dataset.viewing = "false";
    const link = list.querySelector('a[href="#' + firstId + '"]');
    if (!link) return;
    link.dataset.viewing = "true";
  }

  // get first or previous toc listed element in top half of view
  function getFirstInView(query) {
    // get all elements matching query and with id
    const elements = document.querySelectorAll(query);
    const elementsWithIds = [];
    for (const element of elements) {
      if (element.id) elementsWithIds.push(element);
    }

    // get first or previous element in top half of view
    for (let i = 0; i < elementsWithIds.length; i++) {
      const element = elementsWithIds[i];
      const prevElement = elementsWithIds[Math.max(0, i - 1)];
      if (element.getBoundingClientRect().top >= 0) {
        if (element.getBoundingClientRect().top < window.innerHeight / 2)
          return element.id;
        else return prevElement.id;
      }
    }
  }

  // make panel
  function makePanel() {
    // create panel
    const panel = document.createElement("div");
    panel.id = "toc_panel";
    if (bullets === "true") panel.dataset.bullets = "true";

    // create header
    const header = document.createElement("div");
    header.id = "toc_header";

    // create toc button
    const button = document.createElement("button");
    button.id = "toc_button";
    button.innerHTML = document.querySelector(".icon_th_list").innerHTML;
    button.title = "Table of Contents";
    button.classList.add("icon_button");

    // create header text
    const text = document.createElement("h4");
    text.innerHTML = "Table of Contents";

    // create container for toc list
    const list = document.createElement("div");
    list.id = "toc_list";

    // attach click listeners
    panel.addEventListener("click", onPanelClick);
    header.addEventListener("click", onHeaderClick);
    button.addEventListener("click", onButtonClick);

    // attach elements
    header.appendChild(button);
    header.appendChild(text);
    panel.appendChild(header);
    panel.appendChild(list);

    return panel;
  }

  // create toc entries (links) to each element of the specified types
  function makeEntries(panel) {
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) {
      // do not add link if element doesn't have assigned id
      if (!element.id) continue;

      // create link/list item
      const link = document.createElement("a");
      link.classList.add("toc_link");
      switch (element.tagName.toLowerCase()) {
        case "h1":
          link.dataset.level = "1";
          break;
        case "h2":
          link.dataset.level = "2";
          break;
        case "h3":
          link.dataset.level = "3";
          break;
        case "h4":
          link.dataset.level = "4";
          break;
      }
      link.title = element.innerText;
      let text = element.innerText;
      if (text.length > charLimit) text = text.slice(0, charLimit) + "...";
      link.innerHTML = text;
      link.href = "#" + element.id;
      link.addEventListener("click", onLinkClick);

      // attach link
      panel.querySelector("#toc_list").appendChild(link);
    }
  }

  // when panel is clicked
  function onPanelClick(event) {
    // stop click from propagating to window/document and closing panel
    event.stopPropagation();
  }

  // when header itself is clicked
  function onHeaderClick(event) {
    togglePanel();
  }

  // when button is clicked
  function onButtonClick(event) {
    togglePanel();
    // stop header underneath button from also being clicked
    event.stopPropagation();
  }

  // when link is clicked
  function onLinkClick(event) {
    if (clickClose === "true" || (clickClose === "auto" && isSmallScreen()))
      closePanel();
    else openPanel();
  }

  // open panel if closed, close if opened
  function togglePanel() {
    const panel = document.getElementById("toc_panel");
    if (!panel) return;

    if (panel.dataset.open === "true") closePanel();
    else openPanel();
  }

  // open panel
  function openPanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "true";
  }

  // close panel
  function closePanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "false";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- th list icon -->

<template class="icon_th_list">
  <!-- modified from: https://fontawesome.com/icons/th-list -->
  <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
    <path
      fill="currentColor"
      d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* toc panel */
    #toc_panel {
      box-sizing: border-box;
      position: fixed;
      top: 0;
      left: 0;
      background: #ffffff;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      z-index: 2;
    }

    /* toc panel when closed */
    #toc_panel[data-open="false"] {
      min-width: 60px;
      width: 60px;
      height: 60px;
      border-right: solid 1px #bdbdbd;
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc panel when open */
    #toc_panel[data-open="true"] {
      min-width: 260px;
      max-width: 480px;
      /* keep panel edge consistent distance away from "page" edge */
      width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
      bottom: 0;
      border-right: solid 1px #bdbdbd;
    }

    /* toc panel header */
    #toc_header {
      box-sizing: border-box;
      display: flex;
      flex-direction: row;
      align-items: center;
      height: 60px;
      margin: 0;
      padding: 20px;
    }

    /* toc panel header when hovered */
    #toc_header:hover {
      cursor: pointer;
    }

    /* toc panel header when panel open */
    #toc_panel[data-open="true"] > #toc_header {
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc open/close header button */
    #toc_button {
      margin-right: 20px;
    }

    /* hide toc list and header text when closed */
    #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
    #toc_panel[data-open="false"] > #toc_list {
      display: none;
    }

    /* toc list of entries */
    #toc_list {
      box-sizing: border-box;
      width: 100%;
      padding: 20px;
      position: absolute;
      top: calc(60px + 1px);
      bottom: 0;
      overflow: auto;
    }

    /* toc entry, link to section in document */
    .toc_link {
      display: block;
      padding: 5px;
      position: relative;
      font-weight: 600;
      text-decoration: none;
    }

    /* toc entry when hovered or when "viewed" */
    .toc_link:hover,
    .toc_link[data-viewing="true"] {
      background: #f5f5f5;
    }

    /* toc entry, level 1 indentation */
    .toc_link[data-level="1"] {
      margin-left: 0;
    }

    /* toc entry, level 2 indentation */
    .toc_link[data-level="2"] {
      margin-left: 20px;
    }

    /* toc entry, level 3 indentation */
    .toc_link[data-level="3"] {
      margin-left: 40px;
    }

    /* toc entry, level 4 indentation */
    .toc_link[data-level="4"] {
      margin-left: 60px;
    }

    /* toc entry bullets */
    #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
      position: absolute;
      left: -15px;
      top: -1px;
      font-size: 1.5em;
    }

    /* toc entry, level 2 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
      content: "\2022";
    }

    /* toc entry, level 3 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
      content: "\25AB";
    }

    /* toc entry, level 4 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
      content: "-";
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* push <body> ("page") element down to make room for toc icon */
    .toc_body_nudge {
      padding-top: 60px;
    }

    /* toc icon when panel closed and not hovered */
    #toc_panel[data-open="false"]:not(:hover) {
      background: rgba(255, 255, 255, 0.75);
    }
  }

  /* always hide toc panel on print */
  @media only print {
    #toc_panel {
      display: none;
    }
  }
</style>
<!-- 
  Tooltips Plugin

  Makes it such that when the user hovers or focuses a link to a citation or
  figure, a tooltip appears with a preview of the reference content, along with
  arrows to navigate between instances of the same reference in the document.
-->

<script type="module">
  // whether user must click off to close tooltip instead of just un-hovering
  const clickClose = "false";
  // delay (in ms) between opening and closing tooltip
  const delay = "100";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach hover and focus listeners to link
      link.addEventListener("mouseover", onLinkHover);
      link.addEventListener("mouseleave", onLinkUnhover);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("touchend", onLinkTouch);
    }

    // attach mouse, key, and resize listeners to window
    window.addEventListener("mousedown", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("keyup", onKeyUp);
    window.addEventListener("resize", onResize);
  }

  // when link is hovered
  function onLinkHover() {
    // function to open tooltip
    const delayOpenTooltip = function () {
      openTooltip(this);
    }.bind(this);

    // run open function after delay
    this.openTooltipTimer = window.setTimeout(delayOpenTooltip, delay);
  }

  // when mouse leaves link
  function onLinkUnhover() {
    // cancel opening tooltip
    window.clearTimeout(this.openTooltipTimer);

    // don't close on unhover if option specifies
    if (clickClose === "true") return;

    // function to close tooltip
    const delayCloseTooltip = function () {
      // if tooltip open and if mouse isn't over tooltip, close
      const tooltip = document.getElementById("tooltip");
      if (tooltip && !tooltip.matches(":hover")) closeTooltip();
    };

    // run close function after delay
    this.closeTooltipTimer = window.setTimeout(delayCloseTooltip, delay);
  }

  // when link is focused (tabbed to)
  function onLinkFocus(event) {
    openTooltip(this);
  }

  // when link is touched on touch screen
  function onLinkTouch(event) {
    // attempt to force hover state on first tap always, and trigger
    // regular link click (and navigation) on second tap
    if (event.target === document.activeElement) event.target.click();
    else {
      document.activeElement.blur();
      event.target.focus();
    }
    if (event.cancelable) event.preventDefault();
    event.stopPropagation();
    return false;
  }

  // when mouse is clicked anywhere in window
  function onClick(event) {
    closeTooltip();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("tooltip_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("tooltip_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeTooltip();
        break;
    }
  }

  // when window is resized or zoomed
  function onResize() {
    closeTooltip();
  }

  // get all links of types we wish to handle
  function getLinks() {
    const queries = [];
    // exclude buttons, anchor links, toc links, etc
    const exclude =
      ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    queries.push('a[href^="#ref-"]' + exclude); // citation links
    queries.push('a[href^="#fig:"]' + exclude); // figure links
    const query = queries.join(", ");
    return document.querySelectorAll(query);
  }

  // get links with same target, get index of link in set, get total
  // same links
  function getSameLinks(link) {
    const sameLinks = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        sameLinks.push(otherLink);
    }

    return {
      elements: sameLinks,
      index: sameLinks.indexOf(link),
      total: sameLinks.length,
    };
  }

  // open tooltip
  function openTooltip(link) {
    // delete tooltip if it exists, start fresh
    closeTooltip();

    // make tooltip element
    const tooltip = makeTooltip(link);

    // if source couldn't be found and tooltip not made, exit
    if (!tooltip) return;

    // make navbar elements
    const navBar = makeNavBar(link);
    if (navBar) tooltip.firstElementChild.appendChild(navBar);

    // attach tooltip to page
    document.body.appendChild(tooltip);

    // position tooltip
    const position = function () {
      positionTooltip(link);
    };
    position();

    // if tooltip contains images, position again after they've loaded
    const imgs = tooltip.querySelectorAll("img");
    for (const img of imgs) img.addEventListener("load", position);
  }

  // close (delete) tooltip
  function closeTooltip() {
    const tooltip = document.getElementById("tooltip");
    if (tooltip) tooltip.remove();
  }

  // make tooltip
  function makeTooltip(link) {
    // get target element that link points to
    const source = getSource(link);

    // if source can't be found, exit
    if (!source) return;

    // create new tooltip
    const tooltip = document.createElement("div");
    tooltip.id = "tooltip";
    const tooltipContent = document.createElement("div");
    tooltipContent.id = "tooltip_content";
    tooltip.appendChild(tooltipContent);

    // make copy of source node and put in tooltip
    const sourceCopy = makeCopy(source);
    tooltipContent.appendChild(sourceCopy);

    // attach mouse event listeners
    tooltip.addEventListener("click", onTooltipClick);
    tooltip.addEventListener("mousedown", onTooltipClick);
    tooltip.addEventListener("touchstart", onTooltipClick);
    tooltip.addEventListener("mouseleave", onTooltipUnhover);

    // (for interaction with lightbox plugin)
    // transfer click on tooltip copied img to original img
    const sourceImg = source.querySelector("img");
    const sourceCopyImg = sourceCopy.querySelector("img");
    if (sourceImg && sourceCopyImg) {
      const clickImg = function () {
        sourceImg.click();
        closeTooltip();
      };
      sourceCopyImg.addEventListener("click", clickImg);
    }

    return tooltip;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
      "class",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // when tooltip is clicked
  function onTooltipClick(event) {
    // when user clicks on tooltip, stop click from transferring
    // outside of tooltip (eg, click off to close tooltip, or eg click
    // off to unhighlight same refs)
    event.stopPropagation();
  }

  // when tooltip is unhovered
  function onTooltipUnhover(event) {
    if (clickClose === "true") return;

    // make sure new mouse/touch/focus no longer over tooltip or any
    // element within it
    const tooltip = document.getElementById("tooltip");
    if (!tooltip) return;
    if (this.contains(event.relatedTarget)) return;

    closeTooltip();
  }

  // make nav bar to go betwen prev/next instances of same reference
  function makeNavBar(link) {
    // find other links to the same source
    const sameLinks = getSameLinks(link);

    // don't show nav bar when singular reference
    if (sameLinks.total <= 1) return;

    // find prev/next links with same target
    const prevLink = getPrevLink(link, sameLinks);
    const nextLink = getNextLink(link, sameLinks);

    // create nav bar
    const navBar = document.createElement("div");
    navBar.id = "tooltip_nav_bar";
    const text = sameLinks.index + 1 + " of " + sameLinks.total;

    // create nav bar prev/next buttons
    const prevButton = document.createElement("button");
    const nextButton = document.createElement("button");
    prevButton.id = "tooltip_prev_button";
    nextButton.id = "tooltip_next_button";
    prevButton.title =
      "Jump to the previous occurence of this item in the document [←]";
    nextButton.title =
      "Jump to the next occurence of this item in the document [→]";
    prevButton.classList.add("icon_button");
    nextButton.classList.add("icon_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;
    nextButton.innerHTML =
      document.querySelector(".icon_caret_right").innerHTML;
    navBar.appendChild(prevButton);
    navBar.appendChild(document.createTextNode(text));
    navBar.appendChild(nextButton);

    // attach click listeners to buttons
    prevButton.addEventListener("click", function () {
      onPrevNextClick(link, prevLink);
    });
    nextButton.addEventListener("click", function () {
      onPrevNextClick(link, nextLink);
    });

    return navBar;
  }

  // get previous link with same target
  function getPrevLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if < 1
    let index;
    if (sameLinks.index - 1 >= 0) index = sameLinks.index - 1;
    else index = sameLinks.total - 1;
    return sameLinks.elements[index];
  }

  // get next link with same target
  function getNextLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if > total
    let index;
    if (sameLinks.index + 1 <= sameLinks.total - 1) index = sameLinks.index + 1;
    else index = 0;
    return sameLinks.elements[index];
  }

  // get element that is target of link or url hash
  function getSource(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector('[id="' + id + '"]');
    if (!target) return;

    // if ref or figure, modify target to get expected element
    if (id.indexOf("ref-") === 0) target = target.querySelector(":nth-child(2)");
    else if (id.indexOf("fig:") === 0) target = target.querySelector("figure");

    return target;
  }

  // when prev/next arrow button is clicked
  function onPrevNextClick(link, prevNextLink) {
    if (link && prevNextLink)
      goToElement(prevNextLink, window.innerHeight * 0.5);
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);
    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // determine position to place tooltip based on link position in
  // viewport and tooltip size
  function positionTooltip(link, left, top) {
    const tooltipElement = document.getElementById("tooltip");
    if (!tooltipElement) return;

    // get convenient vars for position/dimensions of
    // link/tooltip/page/view
    link = getRectInPage(link);
    const tooltip = getRectInPage(tooltipElement);
    const view = getRectInPage();

    // horizontal positioning
    if (left)
      // use explicit value
      left = left;
    else if (link.left + tooltip.width < view.right)
      // fit tooltip to right of link
      left = link.left;
    else if (link.right - tooltip.width > view.left)
      // fit tooltip to left of link
      left = link.right - tooltip.width;
    // center tooltip in view
    else left = (view.right - view.left) / 2 - tooltip.width / 2;

    // vertical positioning
    if (top)
      // use explicit value
      top = top;
    else if (link.top - tooltip.height > view.top)
      // fit tooltip above link
      top = link.top - tooltip.height;
    else if (link.bottom + tooltip.height < view.bottom)
      // fit tooltip below link
      top = link.bottom;
    else {
      // center tooltip in view
      top = view.top + view.height / 2 - tooltip.height / 2;
      // nudge off of link to left/right if possible
      if (link.right + tooltip.width < view.right) left = link.right;
      else if (link.left - tooltip.width > view.left)
        left = link.left - tooltip.width;
    }

    tooltipElement.style.left = left + "px";
    tooltipElement.style.top = top + "px";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* tooltip container */
    #tooltip {
      position: absolute;
      width: 50%;
      min-width: 240px;
      max-width: 75%;
      z-index: 1;
    }

    /* tooltip content */
    #tooltip_content {
      margin-bottom: 5px;
      padding: 20px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
      overflow-wrap: break-word;
    }

    /* tooltip copy of paragraphs and figures */
    #tooltip_content > p,
    #tooltip_content > figure {
      margin: 0;
      max-height: 320px;
      overflow-y: auto;
    }

    /* tooltip copy of <img> */
    #tooltip_content > figure > img,
    #tooltip_content > figure > svg {
      max-height: 260px;
    }

    /* navigation bar */
    #tooltip_nav_bar {
      margin-top: 10px;
      text-align: center;
    }

    /* navigation bar previous/next buton */
    #tooltip_nav_bar > .icon_button {
      position: relative;
      top: 3px;
    }

    /* navigation bar previous button */
    #tooltip_nav_bar > .icon_button:first-of-type {
      margin-right: 5px;
    }

    /* navigation bar next button */
    #tooltip_nav_bar > .icon_button:last-of-type {
      margin-left: 5px;
    }
  }

  /* always hide tooltip on print */
  @media only print {
    #tooltip {
      display: none;
    }
  }
</style>
<!--
  Analytics Plugin (third-party) 
  
  Copy and paste code from Google Analytics or similar service here.
-->
<!-- 
  Annotations Plugin

  Allows public annotation of the  manuscript. See https://web.hypothes.is/.
-->

<script type="module">
  // configuration
  window.hypothesisConfig = function () {
    return {
      branding: {
        accentColor: "#2196f3",
        appBackgroundColor: "#f8f8f8",
        ctaBackgroundColor: "#f8f8f8",
        ctaTextColor: "#000000",
        selectionFontFamily: "Open Sans, Helvetica, sans serif",
        annotationFontFamily: "Open Sans, Helvetica, sans serif",
      },
    };
  };

  // hypothesis client script
  const embed = "https://hypothes.is/embed.js";
  // hypothesis annotation count query url
  const query = "https://api.hypothes.is/api/search?limit=0&url=";

  // start script
  function start() {
    const button = makeButton();
    document.body.insertBefore(button, document.body.firstChild);
    insertCount(button);
  }

  // make button
  function makeButton() {
    // create button
    const button = document.createElement("button");
    button.id = "hypothesis_button";
    button.innerHTML = document.querySelector(".icon_hypothesis").innerHTML;
    button.title = "Hypothesis annotations";
    button.classList.add("icon_button");

    function onClick(event) {
      onButtonClick(event, button);
    }

    // attach click listeners
    button.addEventListener("click", onClick);

    return button;
  }

  // insert annotations count
  async function insertCount(button) {
    // get annotation count from Hypothesis based on url
    let count = "-";
    try {
      const canonical = document.querySelector('link[rel="canonical"]');
      const location = window.location;
      const url = encodeURIComponent((canonical || location).href);
      const response = await fetch(query + url);
      const json = await response.json();
      count = json.total || "-";
    } catch (error) {
      console.log(error);
    }

    // put count into button
    const counter = document.createElement("span");
    counter.id = "hypothesis_count";
    counter.innerHTML = count;
    button.title = "View " + count + " Hypothesis annotations";
    button.append(counter);
  }

  // when button is clicked
  function onButtonClick(event, button) {
    const script = document.createElement("script");
    script.src = embed;
    document.body.append(script);
    button.remove();
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- hypothesis icon -->

<template class="icon_hypothesis">
  <!-- modified from: https://simpleicons.org/icons/hypothesis.svg / https://git.io/Jf1VB -->
  <svg width="16" height="16" viewBox="0 0 24 24" tabindex="-1">
    <path
      fill="currentColor"
      d="M3.43 0C2.5 0 1.72 .768 1.72 1.72V18.86C1.72 19.8 2.5 20.57 3.43 20.57H9.38L12 24L14.62 20.57H20.57C21.5 20.57 22.29 19.8 22.29 18.86V1.72C22.29 .77 21.5 0 20.57 0H3.43M5.14 3.43H7.72V9.43S8.58 7.72 10.28 7.72C12 7.72 13.74 8.57 13.74 11.24V17.14H11.16V12C11.16 10.61 10.28 10.07 9.43 10.29C8.57 10.5 7.72 11.41 7.72 13.29V17.14H5.14V3.43M18 13.72C18.95 13.72 19.72 14.5 19.72 15.42A1.71 1.71 0 0 1 18 17.13A1.71 1.71 0 0 1 16.29 15.42C16.29 14.5 17.05 13.71 18 13.71Z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  /* hypothesis activation button */
  #hypothesis_button {
    box-sizing: border-box;
    position: fixed;
    top: 0;
    right: 0;
    width: 60px;
    height: 60px;
    background: #ffffff;
    border-radius: 0;
    border-left: solid 1px #bdbdbd;
    border-bottom: solid 1px #bdbdbd;
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
    z-index: 2;
  }

  /* hypothesis button svg */
  #hypothesis_button > svg {
    position: relative;
    top: -4px;
  }

  /* hypothesis annotation count */
  #hypothesis_count {
    position: absolute;
    left: 0;
    right: 0;
    bottom: 5px;
  }

  /* side panel */
  .annotator-frame {
    width: 280px !important;
  }

  /* match highlight color to rest of theme */
  .annotator-highlights-always-on .annotator-hl {
    background-color: #ffeb3b !important;
  }

  /* match focused color to rest of theme */
  .annotator-hl.annotator-hl-focused {
    background-color: #ff8a65 !important;
  }

  /* match bucket bar color to rest of theme */
  .annotator-bucket-bar {
    background: #f5f5f5 !important;
  }

  /* always hide button, toolbar, and tooltip on print */
  @media only print {
    #hypothesis_button {
      display: none;
    }

    .annotator-frame {
      display: none !important;
    }

    hypothesis-adder {
      display: none !important;
    }
  }
</style>
<!-- 
  Mathjax Plugin (third-party) 

  Allows the proper rendering of math/equations written in LaTeX.
  See https://www.mathjax.org/.
-->

<script type="text/x-mathjax-config">
  // configuration
  MathJax.Hub.Config({
    "CommonHTML": { linebreaks: { automatic: true } },
    "HTML-CSS": { linebreaks: { automatic: true } },
    "SVG": { linebreaks: { automatic: true } },
    "fast-preview": { disabled: true }
  });
</script>

<script
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"
  integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A=="
  crossorigin="anonymous"
></script>

<style>
  /* mathjax containers */
  .math.display > span:not(.MathJax_Preview) {
    /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
    display: flex !important;
    overflow-x: auto !important;
    overflow-y: hidden !important;
    justify-content: center;
    align-items: center;
    margin: 0 !important;
  }

  /* right click menu */
  .MathJax_Menu {
    border-radius: 5px !important;
    border: solid 1px #bdbdbd !important;
    box-shadow: none !important;
  }

  /* equation auto-number */
  span[id^="eq:"] > span.math.display + span {
    font-weight: 600;
  }

  /* equation */
  span[id^="eq:"] > span.math.display > span {
    /* nudge to make room for equation auto-number and anchor */
    margin-right: 60px !important;
  }
</style>
</body>
</html>
